
```{r eval=FALSE}
library(Seurat)
library(Signac)
library(tidyverse)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(ggrepel)
library(ggrastr)
library(hdWGCNA)
theme_set(theme_cowplot())
set.seed(12345)

setwd("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF")
data_dir <- 'processed/'
fig_dir <- 'figures/'

# set the current cell-type (oligos)
cur_celltype <- 'ODC'
fig_dir <- paste0('figures/', cur_celltype, '/'); dir.create(fig_dir)

################################################################################
# load Seurat objects
################################################################################

# load the co-embedding object for the oligo lineage:
seurat_obj <- readRDS('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/ODC_coembed_seurat.rds')

# add umap coords:
seurat_obj@meta.data$UMAP_1 <- seurat_obj@reductions$umap@cell.embeddings[,1]
seurat_obj@meta.data$UMAP_2 <- seurat_obj@reductions$umap@cell.embeddings[,2]

# load the ATAC & RNA objects separately, and subset just by the oligo lineage cells:
seurat_atac <- readRDS('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/snATAC_processed_seurat.rds')
DefaultAssay(seurat_atac) <- 'peaks'
seurat_atac$Time.point <- seurat_atac$Timepoint

# load RNA object
seurat_rna <- readRDS("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/snRNA_processed_seurat.rds")

# load old & young t-degs
o_tdegs <- read.csv('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/ODC_Old_tDEGs.csv')
rownames(o_tdegs) <- o_tdegs$gene_id

y_tdegs <- read.csv('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/ODC_Young_tDEGs.csv')
rownames(y_tdegs) <- y_tdegs$gene_id

################################################################################
# subset old/young:
################################################################################

cur_group <- 'Young'
cur_group <- 'Old'

# re-load the trajectory information
load(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_expression_trajectory_data.rda'))

subset.cells <- ((seurat_obj$Time.point == 'Naive' & seurat_obj$subcluster_name == 'OPC') |
(seurat_obj$Time.point == 'Day.5' & seurat_obj$subcluster_name %in% c('OPC', 'NFOL')) |
(seurat_obj$Time.point == 'Day.14' & seurat_obj$subcluster_name %in% c('MF-ODC', 'Int-ODC')) |
(seurat_obj$Time.point == 'Day.30' & seurat_obj$subcluster_name == 'Mat-ODC')) &
(seurat_obj$Group == cur_group)
table(subset.cells)

# subset the seurat obj
seurat_subset <- seurat_obj[,subset.cells]

# cut pseudotime into bins:
seurat_subset <- hdWGCNA:::BinPseudotime(seurat_subset, n_bins=c(10,50,100))

# re-load gl-cCRE table:
link_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_gl-cCREs.csv'))

peak1_ranges <- Signac::StringToGRanges(link_df$Peak1, sep=c('-', '-'))
peak1_ranges$peak <- link_df$Peak1

peak2_ranges <- Signac::StringToGRanges(link_df$Peak2, sep=c('-', '-'))
peak2_ranges$peak <- link_df$Peak2

```


Load RVAE outputs, rank genes by their pseudotime expression patterns

```{r eval=FALSE}

library(future.apply)
library(viridis)

# RVAE directory
outdir <- '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
cur_name <- paste0(cur_celltype, '_', cur_group, '_RVAE')
rvagene_csv <- paste0(outdir, 'reconstructed/',cur_name,'_recon.csv')

# load reconstructed data and z embedding
dataset <- cur_name
recon_df <- read.table(paste0(outdir, 'reconstructed/',cur_name,'_recon.csv'), sep=',', row.names=1, header=TRUE)
z_df <- recon_df[,(ncol(recon_df)-1):ncol(recon_df)]
recon_df <- recon_df[,1:(ncol(recon_df)-2)]
z_df$gene <- rownames(z_df)

scaled <- lapply(rownames(recon_df), range01)
scaled <- do.call(rbind, scaled)
rownames(scaled) <- rownames(recon_df)

# order rows by the time they reach 0.90% max expression
exp_thresh <- 0.90
ordering <- future_lapply(1:nrow(scaled), function(i){
  match(names(scaled[i,])[scaled[i,] >= exp_thresh][1], colnames(scaled))
})
ordering <- unlist(ordering)
ordered <- scaled[rownames(scaled)[order(ordering)],]

ordered_bins <- ordering[order(ordering)]

z_df_tmp <- z_df

ix <- match(rownames(z_df), rownames(ordered))
z_df$rank <- ix
z_df$bin <- ordered_bins[ix]
z_df <- dplyr::arrange(z_df, rank)
z_df$gene <- factor(as.character(z_df$gene), levels=as.character(z_df$gene))


```

Setup motif data

```{r eval=FALSE}

# get list of TFs from atac data
DefaultAssay(seurat_atac) <- 'peaks'
chromvar_motifs <- Motifs(seurat_atac)@motif.names %>%
  unlist %>% as.character
print(length(chromvar_motifs))

# set up motif df
motif_df <- data.frame(
  motif_id = names(Motifs(seurat_atac)@motif.names),
  motif_name_orig = as.character(Motifs(seurat_atac)@motif.names)
)

tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
tmp_ids <- names(Motifs(seurat_atac)@motif.names[chromvar_motifs %in% tmp])

motif_df$motif_name <- gsub("\\s*\\([^\\)]+\\)","", motif_df$motif_name_orig)
tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
motif_df <- subset(motif_df, !(motif_name %in% tmp))

tmp2 <- strsplit(tmp, '::')
names(tmp2) <- tmp
tmp <- do.call(rbind, lapply(1:length(tmp2), function(i){
  data.frame(motif_id = tmp_ids[i], motif_name_orig = names(tmp2)[i], motif_name = unlist(as.character(tmp2[[i]])))
}))
motif_df <- rbind(motif_df, tmp)

# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))

ix <- match(motif_df$motif_name, hg38_mm10_genes$hg38_name)
motif_df$motif_name_mouse <- hg38_mm10_genes$mm10_name[ix]

chromvar_motifs <- unique(motif_df$motif_name_mouse)

# which do we also have RNA data for?
chromvar_motifs <- chromvar_motifs[chromvar_motifs %in% rownames(seurat_rna)]

tf_z_df <- subset(z_df, gene %in% chromvar_motifs)
tf_z_df$motif_id <- motif_df$motif_id[match(tf_z_df$gene, motif_df$motif_name)]

```

Set up a table of all the potential TF-gene links based on TF motifs present in promoter regions and in promoter-linked enhancers 

```{r eval=FALSE}

tfs <- as.character(tf_z_df$gene)

target_df <- data.frame()
for(cur_tf in tfs){
  print(cur_tf)

  cur_tf_id <- motif_df %>% subset(motif_name_mouse == cur_tf) %>% .$motif_id
  cur_bin <- tf_z_df %>% subset(gene == cur_tf) %>% .$bin

  if(length(cur_tf_id) > 1){
    cur_tf_id <- cur_tf_id[1]
  }

  # get sites for this motif:
  cur_motif_sites <- Motifs(seurat_atac)@data[,cur_tf_id]
  cur_motif_sites <- names(cur_motif_sites)[cur_motif_sites == 1]

  # which of these sites overlap a gene promoter?
  promoter_sites <- peakset[peakset$peak %in% cur_motif_sites & peakset$peakType == 'Promoter',]
  promoter_genes <- unique(promoter_sites$nearestGene)
  promoter_genes <- promoter_genes[promoter_genes %in% z_df$gene]

  # which of these sites overlap an enhancer?
  enhancer_links <- subset(link_df, Peak2 %in% cur_motif_sites & Peak2 %in% accessible_peaks)
  enhancer_genes <- unique(enhancer_links$Peak1_nearestGene)
  enhancer_genes <- enhancer_genes[enhancer_genes %in% z_df$gene]

  target_genes <- unique(c(promoter_genes, enhancer_genes))

  cur_target_df <- data.frame(
    source = cur_tf,
    target = target_genes
  )

  cur_target_df$bs_group <- ifelse(
    cur_target_df$target %in% promoter_genes & cur_target_df$target %in% enhancer_genes, 'Both',
    ifelse(cur_target_df$target %in% promoter_genes, "Promoter",
    ifelse(cur_target_df$target %in% enhancer_genes, "Enhancer", NA
  )))

  cur_target_df$target_type <- ifelse(
    cur_target_df$target %in% tfs, 'TF', 'other'
  )

  # add correlation info
  cur_target_df$cor <- as.numeric(tf_cor_mat[cur_tf,as.character(cur_target_df$target)])

  # add Z coordinates of source and target
  ix <- match(cur_target_df$source, z_df$gene)
  cur_target_df$Z1_source <- z_df$Z1[ix]
  cur_target_df$Z2_source <- z_df$Z2[ix]
  ix <- match(cur_target_df$target, z_df$gene)
  cur_target_df$Z1_target <- z_df$Z1[ix]
  cur_target_df$Z2_target <- z_df$Z2[ix]

  # compute distances
  target_coords <- cur_target_df %>% dplyr::select(c(Z1_target, Z2_target))
  rownames(target_coords) <- cur_target_df$target; colnames(target_coords) <- c('Z1', 'Z2')
  source_coords <- cur_target_df[1,c('Z1_source', 'Z2_source')]; colnames(source_coords) <- c('Z1', 'Z2'); rownames(source_coords) <- cur_tf

  if(cur_tf %in% rownames(target_coords)){
    cur_coord <- which(rownames(target_coords) == cur_tf)
    coords <- rbind(target_coords[cur_coord,], target_coords[-cur_coord,])
  } else{
    coords <- rbind(source_coords, target_coords)
  }

  cur_distances <- as.matrix(dist(as.matrix(coords), method='euclidean'))[cur_tf,]
  ix <- match(cur_target_df$target, names(cur_distances))
  cur_target_df$dist <- as.numeric(cur_distances[ix])

  # update whole target df
  target_df <- rbind(target_df, cur_target_df)

}

# write the output
write.csv(target_df, file=paste0(data_dir, cur_celltype, '_', cur_group, '_tf_target_df.csv') )

```

Identify most likely tfs for each gene using XGBoost

```{r eval=FALSE}

library(xgboost)
library(tictoc)

target_df <- read.csv(file=paste0(data_dir, cur_celltype, '_', cur_group, '_tf_target_df.csv') )

# list of target genes to run XGBoost on:
gene_list <- unique(target_df$target)

# set up XGBoost model parameters
model_params <- list(
  objective = 'reg:squarederror',
  max_depth = 3,
  eta = 0.1,
  nthread=16,
  alpha=0.5
)
nfold=5

importance_df <- data.frame()
eval_df <- data.frame()

for(cur_target in gene_list){

  # build the model for one TF and one target gene
  cur_tfs <- target_df %>% subset(target == cur_target) %>% .$source %>% as.character %>% unique
  cur_tfs <- cur_tfs[cur_tfs != cur_target]

  if(length(cur_tfs) < 3){
    counter <- counter + 1
    next
  }

  dat = t(GetAssayData(seurat_rna, slot='data')[c(cur_target, cur_tfs),])

  # model matrix
  x_vars <- dat[,-1]
  y_var <- as.numeric(dat[,1])

  if(all(y_var == 0)){
    print(paste0('skipping ', cur_target))
    next
  }
  xgb <- xgb.cv(
    params = model_params,
    data = x_vars,
    label = y_var,
    nrounds = 100,
    showsd = FALSE,
    nfold = nfold,
    callbacks = list(cb.cv.predict(save_models=TRUE)),
    verbose=FALSE
  )

  # get the CV evaluation info
  xgb_eval <- as.data.frame(xgb$evaluation_log)
  xgb_eval$variable <- cur_target

  # average the importance score from each fold
  importance <- Reduce('+', lapply(1:nfold, function(i){
    cur_imp <- xgb.importance(feature_names = colnames(x_vars), model = xgb$models[[i]])
    ix <- match(colnames(x_vars),  as.character(cur_imp$Feature))
    cur_imp <- as.matrix(cur_imp[ix,-1])
    cur_imp[is.na(cur_imp)] <- 0
    cur_imp
  })) / nfold
  importance <- as.data.frame(importance)

  # add tf and source info
  importance$source <- colnames(x_vars)
  importance$target <- cur_target

  # add pearson correlation of gene expression
  importance$exp_cor <- cor(as.matrix(x_vars), y_var)

  # re-order columns, and re-order rows by gain:
  importance <- importance %>% dplyr::select(c(source, target, Gain, Cover, Frequency, exp_cor))
  importance <- arrange(importance, -Gain)

  # append
  importance_df <- rbind(importance_df, importance)
  eval_df <- rbind(eval_df, xgb_eval)

}

write.csv(importance_df, file=paste0(data_dir, cur_celltype, '_', cur_group, '_xgboost_importance.csv') )

```

# Regulons?