
```{r eval=FALSE}

library(Seurat)
library(Signac)
library(tidyverse)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(ggrepel)

library(ggrastr)

library(hdWGCNA)

theme_set(theme_cowplot())
set.seed(12345)

setwd("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF")
data_dir <- 'processed/'
fig_dir <- 'figures/'


# set the current cell-type (oligos)
cur_celltype <- 'ODC'
fig_dir <- paste0('figures/', cur_celltype, '/'); dir.create(fig_dir)


################################################################################
# load Seurat objects
################################################################################

# human MS datasets
seurat_ab <- readRDS(file='/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/Absinta_2021.rds')
seurat_sh <- readRDS(file='/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/Schirmer_2019.rds')

# load the co-embedding object for the oligo lineage:
seurat_obj <- readRDS('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/ODC_coembed_seurat.rds')

# load the ATAC & RNA objects separately, and subset just by the oligo lineage cells:
seurat_atac <- readRDS('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/signac_final_07-02-21_update.rds')
DefaultAssay(seurat_atac) <- 'peaks'
seurat_atac$Time.point <- seurat_atac$Timepoint

seurat_rna <- readRDS("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/seurat_objects/NucSeq_final_doublets_removed_online_inmf_seurat.rds")

atac_ix <- na.omit(match(colnames(seurat_obj), colnames(seurat_atac)))
rna_ix <- na.omit(match(colnames(seurat_obj), colnames(seurat_rna)))

# subset by atac/rna cells that are in the oligo pseudotime object:
seurat_atac_full <- seurat_atac
seurat_rna_full <- seurat_rna

seurat_atac <- seurat_atac_full[,atac_ix]
seurat_rna <- seurat_rna_full[,rna_ix]


# load old & young t-degs
o_tdegs <- read.csv('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/ODC_Old_tDEGs.csv')
rownames(o_tdegs) <- o_tdegs$gene_id

y_tdegs <- read.csv('/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/ODC_Young_tDEGs.csv')
rownames(y_tdegs) <- y_tdegs$gene_id


#################################################################################
# Load the results from the TF net analysis
################################################################################

old_target_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', 'Old_tf_target_df.csv'))
young_target_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_',  'Young_tf_target_df.csv'))

old_importance_df <-  read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', 'Old_xgboost_importance.csv') )
young_importance_df <-  read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', 'Young_xgboost_importance.csv') )

#################################################################################
# Load the human DEGs 
################################################################################

filedir <- "/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/DEGs/human/"
deg_files <- dir(filedir)
human_degs  <- data.frame()
for(i in 1:length(deg_files)){

   deg_file <- deg_files[i]
  if(grepl('TF', deg_file)){
    next
  } 
  name <- gsub('_degs.csv', '',deg_file)
  print(name)
  degs <- read.csv(paste0(filedir,deg_file))
  degs$comparison <- name
  human_degs <- rbind(human_degs, degs)
}

#################################################################################
# Load the mouse young v old DEGs
################################################################################

# need to copy from HPC3
filedir <- "/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/DEGs/mouse/"
timepoints <- c('Naive', 'Day5', 'Day14', 'Day30')

mouse_degs <- data.frame()
for(tp in timepoints){
  
  print(tp)
  degs <- read.csv(paste0(filedir, tp, '_DEGs.csv'))
  degs$timepoint <- tp 
  mouse_degs <- rbind(mouse_degs, degs)

}

#################################################################################
# Load hdWGCNA outputs:
################################################################################

seurat_hd <- readRDS("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/hdWGCNA/data/oligo_snRNA_hdWGCNA.rds")


```

Setup motif data (important for later steps)

```{r eval=FALSE}

load(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_Old_expression_trajectory_data.rda'))
#old_switches <- switch_genes
old_ordered <- ordered

load(file=paste0(data_dir, cur_celltype, '_Young_expression_trajectory_data.rda'))
#young_switches <- switch_genes
young_ordered <- ordered


# get list of TFs from atac data
DefaultAssay(seurat_atac) <- 'peaks'
chromvar_motifs <- Motifs(seurat_atac)@motif.names %>%
  unlist %>% as.character
print(length(chromvar_motifs))

# set up motif df
motif_df <- data.frame(
  motif_id = names(Motifs(seurat_atac)@motif.names),
  motif_name_orig = as.character(Motifs(seurat_atac)@motif.names)
)

# 1. split apart cases where two genes are one motif delimited by ::
# 2. remove "()" characters and anything within them

tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
tmp_ids <- names(Motifs(seurat_atac)@motif.names[chromvar_motifs %in% tmp])

motif_df$motif_name <- gsub("\\s*\\([^\\)]+\\)","", motif_df$motif_name_orig)
tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
motif_df <- subset(motif_df, !(motif_name %in% tmp))

tmp2 <- strsplit(tmp, '::')
names(tmp2) <- tmp
tmp <- do.call(rbind, lapply(1:length(tmp2), function(i){
  data.frame(motif_id = tmp_ids[i], motif_name_orig = names(tmp2)[i], motif_name = unlist(as.character(tmp2[[i]])))
}))
motif_df <- rbind(motif_df, tmp)

# convert gene names to mouse style by matching genes in ensembl
# motif_df$motif_name <- str_to_title(motif_df$motif_name) # old

# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))

ix <- match(motif_df$motif_name, hg38_mm10_genes$hg38_name)
motif_df$motif_name_mouse <- hg38_mm10_genes$mm10_name[ix]


chromvar_motifs <- unique(motif_df$motif_name_mouse)

# which do we also have RNA data for?
chromvar_motifs <- chromvar_motifs[chromvar_motifs %in% rownames(seurat_rna)]

tf_z_df <- subset(z_df, gene %in% chromvar_motifs)
tf_z_df$motif_id <- motif_df$motif_id[match(tf_z_df$gene, motif_df$motif_name_mouse)]


```

Get Switch genes for young & old 

```{r eval=FALSE}

# switch == 80% of max expression in only 2 bins 
exp_thresh <- 0.8
switch_thresh <- 2

load(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', 'Young', '_expression_trajectory_data.rda'))
young_ordered <- ordered

load(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', 'Old', '_expression_trajectory_data.rda'))
old_ordered <- ordered

# number of bins greater than or equal to the thresh
n_bins <- (young_ordered >= exp_thresh) %>% rowSums
young_switch_genes <- names(which(n_bins <= switch_thresh))

# number of bins greater than or equal to the thresh
n_bins <- (old_ordered >= exp_thresh) %>% rowSums
old_switch_genes <- names(which(n_bins <= switch_thresh))

length(young_switch_genes) # 2908
length(old_switch_genes) # 2691

# define the grey genes
modules_oligo <- GetModules(seurat_hd, wgcna_name = 'oligo')
modules_opc <- GetModules(seurat_hd, wgcna_name = 'opc')
grey_genes <- unique(c(
  subset(modules_oligo, module == 'grey') %>% .$gene_name,
  subset(modules_opc, module == 'grey') %>% .$gene_name
))

young_switch_genes <- young_switch_genes[!(young_switch_genes %in% grey_genes)]
old_switch_genes <- old_switch_genes[!(old_switch_genes %in% grey_genes)]

length(intersect(young_switch_genes, old_switch_genes)) # 1033
length(setdiff(young_switch_genes, old_switch_genes)) # 780

switch_genes <- unique(c(young_switch_genes, old_switch_genes))

```


Plot motif logos for the critical TFs
```{r eval=FALSE}

library(JASPAR2020)

# JASPAR 2020
pfm_core <- TFBSTools::getMatrixSet(
  x = JASPAR2020,
  opts = list(collection = "CORE", tax_group = 'vertebrates', all_versions = FALSE)
)

critical_nodes <- c('Sox10', 'Nkx2-2', 'Nkx6-2', 'Tcf4', 'Olig1', 'Olig2')

cur_tf <- critical_nodes[1]

for(cur_tf in critical_nodes){

  cur_id <- subset(motif_df, motif_name_mouse == cur_tf) %>% .$motif_id

  cur_pfm <- as.matrix(pfm_core[[cur_id]])

    p <- ggplot() +
    ggseqlogo::geom_logo( cur_pfm) +
    ggseqlogo::theme_logo() +
    xlab('') + ylab('') + theme(
      axis.text.x=element_blank(),
      axis.text.y=element_blank(),
      axis.title.y = element_text(angle=0),
      plot.margin = margin(t = 0,  # Top margin
                              r = 0,  # Right margin
                              b = 0,  # Bottom margin
                              l = 0) # Left margin
    )

  pdf(paste0(fig_dir, 'motif_logo_', cur_tf, '.pdf'), width=4, height=2)
  print(p)
  dev.off()
  
}

```


Questions for TF prioritization 

1. What TFs are DE between young & old mice?
2. What TFs are DE between MS & control humans?
3. Of those, which are intersecting?
4. Of those, which are trajectory DEGs?
5. Of those, which are in our core TF network from the pathfinding?
6. Which are "switch" genes?

```{r eval=FALSE}

# should we not consider the Naive timepoint?
cur_degs_mouse <- subset(mouse_degs, 
    gene %in% unique(motif_df$motif_name_mouse) & 
    p_val_adj <= 0.05 & 
    abs(avg_log2FC) >= 0.1 & 
    group %in% c('NFOL', 'OPC', 'Int-ODC', 'MF-ODC', 'Mat-ODC') &
    timepoint != 'Naive'
)
dim(cur_degs_mouse)

# considering all of the comparisons
cur_degs_human <- subset(human_degs,
    gene %in% unique(motif_df$motif_name) &
    p_val_adj <= 0.05 &
    abs(avg_log2FC) >= 0.1 &
    group %in% c('OPC', 'ODC')
)
dim(cur_degs_human)

ix <- match(cur_degs_human$gene, motif_df$motif_name)
cur_degs_human$gene_mouse <- motif_df$motif_name_mouse[ix]

# 53 TFs are consistent between mouse and human in this comparison
shared_tfs <- intersect(cur_degs_mouse$gene, cur_degs_human$gene_mouse)

# which of these are also dynamic genes?
dynamic_genes <- unique(c(o_tdegs$gene, y_tdegs$gene))
shared_tfs <- shared_tfs[shared_tfs %in% dynamic_genes] 
length(shared_tfs)

shared_tfs[shared_tfs %in% dynamic_genes] %>% length
shared_tfs[shared_tfs %in% y_tdegs$gene] %>% table

# which of these are "switch" genes?
shared_tfs[shared_tfs %in% dynamic_genes & shared_tfs %in% switch_genes]
shared_tfs[shared_tfs %in% dynamic_genes & shared_tfs %in% switch_genes] %>% length

# which of these are in the TF pathfinding?
shared_tfs[shared_tfs %in% dynamic_genes & shared_tfs %in% pathfinding_tfs$Old]
shared_tfs[shared_tfs %in% dynamic_genes & shared_tfs %in% pathfinding_tfs$Young]


# melt matrix for plotting:ff
tmp1 <- young_ordered[rownames(young_ordered) %in% shared_tfs,]
plot_df <- reshape2::melt(as.matrix( tmp1[rev(rownames(tmp1)),]))

shared_tfs <- rownames(tmp1)


#----------------------------------------------------------------#
# Gene Overlap
#----------------------------------------------------------------#

library(eulerr)
library(GeneOverlap)

mouse_genes <- unique(cur_degs_mouse$gene)
human_genes <- unique(cur_degs_human$gene)

overlap_list <- c(
  'Mouse' = length(mouse_genes),
  'Human' =  length(human_genes),
  'Mouse&Human' =  length(intersect(cur_degs_mouse$gene, cur_degs_human$gene_mouse))
)


pdf(paste0(fig_dir, cur_celltype, '_tfs_prioritized_overlaps.pdf'), width=4, height=4)
plot(euler(overlap_list), quantities = TRUE)
dev.off()


cur_overlap <- testGeneOverlap(newGeneOverlap(
    cur_degs_mouse$gene,
    cur_degs_human$gene_mouse,
    genome.size=nrow(motif_df)
))

print(cur_overlap)

# OR = 16.1 
# p = 1.5e-24

#----------------------------------------------------------------#
# Trjaectory heatmaps
#----------------------------------------------------------------#

p1 <- plot_df %>% 
  ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='seagreen') +
  scale_fill_gradient2(low='white', high='seagreen') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_text(face='italic'),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 

# melt matrix for plotting:ff
tmp2 <- old_ordered[rownames(old_ordered) %in% shared_tfs,]
tmp2 <- tmp2[rownames(tmp1),]
plot_df <- reshape2::melt(as.matrix(tmp2[rev(rownames(tmp2)),]))


p3 <- plot_df %>% 
  ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='darkorchid3') +
  scale_fill_gradient2(low='white', high='darkorchid3') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 

patch <- p1 + p3 +  plot_layout(guides = 'collect')


pdf(paste0(fig_dir, cur_celltype, '_tfs_prioritized2.pdf'), width=6, height=8)
print(patch)
dev.off()


#----------------------------------------------------------------
# Plot trajectory heatmap for regulons !!
#----------------------------------------------------------------

young_regulon_scores <- read.table(
  file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_Young', '_tf_regulon_scores_pos.csv'),
  sep=',', header=TRUE, row.names=1
)
colnames(young_regulon_scores) <- gsub("[.]", "-", colnames(young_regulon_scores))

old_regulon_scores <- read.table(
  file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_Old', '_tf_regulon_scores_pos.csv'),
  sep=',', header=TRUE, row.names=1
)
colnames(old_regulon_scores) <- gsub("[.]", "-", colnames(old_regulon_scores))


# options 
ps_bin_name <- 'pseudotime_bins_50'

cur_bcs <- subset(seurat_rna@meta.data, Group == "Young") %>% rownames()

# which regulon scores to use?
regulon_scores <- young_regulon_scores
regulon_scores <- regulon_scores[cur_bcs,]

# name of the TFs to plot:
cur_tfs <- shared_tfs

# combine the metadata with the TF scores
meta <- seurat_rna@meta.data[cur_bcs,]
plot_df <- cbind(meta, regulon_scores[,cur_tfs])

# compute the average ME in each pseudotime bin
avg_scores <- plot_df %>%
        group_by(get(ps_bin_name)) %>%
        select(all_of(cur_tfs)) %>%
        summarise_all(mean)
for(cur_tf in cur_tfs){
  avg_scores[,cur_tf] <- scale01(avg_scores[,cur_tf])
}

colnames(avg_scores)[1] <- 'bin'
avg_df <- reshape2::melt(avg_scores)
avg_df$bin <- as.numeric(avg_df$bin)
avg_df$type <- 'Regulon'
avg_df$variable <- factor(as.character(avg_df$variable), levels = rev(shared_tfs))
young_avg_df <- avg_df 



cur_bcs <- subset(seurat_rna@meta.data, Group == "Old") %>% rownames()

# which regulon scores to use?
regulon_scores <- old_regulon_scores
regulon_scores <- regulon_scores[cur_bcs,]

# name of the TFs to plot:
cur_tfs <- shared_tfs

# combine the metadata with the TF scores
meta <- seurat_rna@meta.data[cur_bcs,]
plot_df <- cbind(meta, regulon_scores[,cur_tfs])

# compute the average ME in each pseudotime bin
avg_scores <- plot_df %>%
        group_by(get(ps_bin_name)) %>%
        select(all_of(cur_tfs)) %>%
        summarise_all(mean)
for(cur_tf in cur_tfs){
  avg_scores[,cur_tf] <- scale01(avg_scores[,cur_tf])
}

colnames(avg_scores)[1] <- 'bin'
avg_df <- reshape2::melt(avg_scores)
avg_df$bin <- as.numeric(avg_df$bin)
avg_df$type <- 'Regulon'
avg_df$variable <- factor(as.character(avg_df$variable), levels = rev(shared_tfs))
old_avg_df <- avg_df 



p2 <- young_avg_df %>% 
  ggplot(aes(y=variable, x=as.numeric(bin), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='seagreen3') +
  scale_fill_gradient2(low='white', high='seagreen3') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 


p4 <- old_avg_df %>% 
  ggplot(aes(y=variable, x=as.numeric(bin), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='mediumorchid1') +
  scale_fill_gradient2(low='white', high='mediumorchid1') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    #axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 

patch <- (p1 | p2 | p3 | p4) +  plot_layout(guides = 'collect')

pdf(paste0(fig_dir, cur_celltype, '_tfs_prioritized_regulons.pdf'), width=7, height=8)
print(patch)
dev.off()


#----------------------------------------------------------------
# what modules are these in ?
#----------------------------------------------------------------

mod_colors <- dplyr::select(modules_oligo, c(module, color)) %>% dplyr::distinct()
mod_cp <- mod_colors$color; names(mod_cp) <- as.character(mod_colors$module)
mods <- levels(modules_oligo$module)
mod_cp <- mod_cp[mods]

plot_df <- subset(modules_oligo, gene_name %in% shared_tfs) 
plot_df <- plot_df[rev(shared_tfs),]
plot_df$gene_name <- factor(as.character(plot_df$gene_name), levels = rev(shared_tfs))


p <- ggplot(plot_df, aes(y=gene_name, fill=module, x=1)) + 
  geom_tile() + coord_fixed() + 
  geom_text(aes(label=module), size=2) +
  scale_fill_manual(values=mod_cp) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )

pdf(paste0(fig_dir, cur_celltype, '_tfs_prioritized_regulons_mods.pdf'), width=6, height=8)
print(p)
dev.off()


```


Organize a table with the prioritized TF results:

Columns: 
* TF 
* Mouse_DEG_FC 
* Mouse_DEG_pval 
* Mouse_DEG_comparison 
* Human_DEG_FC 
* Human_DEG_pval 
* Human_DEG_comparison 
* Dynamic?
* Switch ?
* Young / Old pathfinding?

```{r eval=FALSE}

selected_tf_df <- Reduce(rbind, lapply(shared_tfs, function(cur_tf){

    # get the mouse & human DEGs for these TFs:
    cur_mouse <- subset(cur_degs_mouse, gene == cur_tf) %>% 
        dplyr::rename(comparison=timepoint) %>% 
        select(-c(pct.1, pct.2))
    cur_human <- subset(cur_degs_human, gene_mouse == cur_tf) %>% 
        mutate(gene = gene_mouse) %>% 
        select(-c(gene_mouse, pct.1, pct.2))
    cur_degs <- rbind(cur_mouse, cur_human)


    cur_degs$dynamic_young <- cur_degs$gene %in% y_tdegs$gene
    cur_degs$dynamic_old <- cur_degs$gene %in% o_tdegs$gene
    cur_degs$switch_young <- cur_degs$gene %in% young_switch_genes
    cur_degs$switch_old <- cur_degs$gene %in% old_switch_genes
    cur_degs$tfnet_young <- cur_degs$gene %in% pathfinding_tfs$Young
    cur_degs$tfnet_old <- cur_degs$gene %in% pathfinding_tfs$Old
    cur_degs

}))

write.table(selected_tf_df, file=paste0(data_dir, 'prioritized_TFs.tsv'), sep='\t', row.names=FALSE, quote=FALSE)

selected_tf_df <- read.table(file=paste0(data_dir, 'prioritized_TFs.tsv'), sep='\t', header=TRUE)


cur_tf %in% o_tdegs$gene
cur_tf %in% y_tdegs$gene

```

Enrichment analysis of switch-like genes, 

plot the trajectory heatmaps for all switch-like genes

```{r eval=FALSE}

#----------------------------------------------------------------#
# enrichment analysis
#----------------------------------------------------------------#

library(enrichR)


dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021', 'WikiPathway_2021_Mouse', 'KEGG_2021_Mouse')


  # setup input list for enrichr loop
  input_list <- list(
    'Young' = setdiff(young_switch_genes, old_switch_genes),
    'Old' = setdiff(old_switch_genes, young_switch_genes),
    'Shared' = intersect(young_switch_genes, old_switch_genes)
  )


# run enrichr and combine outputs
combined_output <- do.call(rbind, lapply(names(input_list), function(x){
  if(length(input_list[[x]]) > 0){
    cur_enrich <- enrichr(input_list[[x]], dbs)
  } else{return(data.frame())}
  cur_df <- do.call(rbind, lapply(dbs, function(cur_db){
    df <- cur_enrich[[cur_db]]
    if(nrow(df) > 1){df$group <- x; df$db <- cur_db;}
    else{df <- data.frame()}
    df
  }))
}))

combined_output$ngenes <- unlist(lapply(strsplit(combined_output$Genes, ';'), function(x){length(x)}))
combined_output <- subset(combined_output, ngenes >= 3)
#combined_output <- subset(combined_output, P.value <= 0.05)


write.table(combined_output, file=paste0(data_dir, cur_celltype,'_switchgene_GO_terms_full.tsv'), quote=FALSE, row.names=FALSE, sep='\t')

combined_output %>% subset(P.value <= 0.05) %>% 
  write.table(file=paste0(data_dir, cur_celltype, '_switchgene_GO_terms_signif.tsv'), quote=FALSE, row.names=FALSE, sep='\t')



combined_output$ngenes <- unlist(lapply(strsplit(combined_output$Genes, ';'), function(x){length(x)}))
combined_output <- subset(combined_output, ngenes >= 3)

write.table(combined_output, file=paste0(data_dir, cur_celltype, '_', cur_group, '_tfnet_target_GO_terms.tsv'), quote=FALSE, row.names=FALSE, sep='\t')

table(combined_output$tf)


selected_terms <- read.delim(paste0(data_dir, 'ODC_switchgenes_GO_selected.txt'), sep='\t', header=1)
selected_terms$Term <- str_replace(selected_terms$Term, " \\s*\\([^\\)]+\\)", "")


p1 <- selected_terms  %>% subset(group == "Old") %>%
  ggplot(aes(x=log(Combined.Score), y=reorder(Term, Combined.Score)))+
  geom_bar(stat='identity', position='identity', fill = "mediumorchid1") +
  geom_text(aes(label=Term), x=.1, color='black', size=3.5, hjust='left') +
  xlab('log(Enrichment)') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) +
  theme(
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    legend.title = element_blank(),
    axis.ticks.y=element_blank(),
    axis.text.y=element_blank(),
    axis.line.y=element_blank(),
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_blank()
  )


p2 <- selected_terms  %>% subset(group == "Shared") %>%
  ggplot(aes(x=log(Combined.Score), y=reorder(Term, Combined.Score)))+
  geom_bar(stat='identity', position='identity', fill = "lightgrey") +
  geom_text(aes(label=Term), x=.1, color='black', size=3.5, hjust='left') +
  xlab('log(Enrichment)') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) +
  theme(
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    legend.title = element_blank(),
    axis.ticks.y=element_blank(),
    axis.text.y=element_blank(),
    axis.line.y=element_blank(),
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_blank()
  )


p3 <- selected_terms  %>% subset(group == "Young") %>%
  ggplot(aes(x=log(Combined.Score), y=reorder(Term, Combined.Score)))+
  geom_bar(stat='identity', position='identity', fill = "seagreen3") +
  geom_text(aes(label=Term), x=.1, color='black', size=3.5, hjust='left') +
  xlab('log(Enrichment)') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) +
  theme(
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    legend.title = element_blank(),
    axis.ticks.y=element_blank(),
    axis.text.y=element_blank(),
    axis.line.y=element_blank(),
    plot.title = element_text(hjust = 0.5),
    axis.title.y = element_blank()
  )


pdf(paste0(fig_dir, 'ODC_switchgenes_GO.pdf'), width= 6, height=2.5 , useDingbats=FALSE)
p3 + p2 + p1 
dev.off()


#----------------------------------------------------------------#
# Plot trajectory heatmaps
#----------------------------------------------------------------#


# setup input list for enrichr loop
input_list <- list(
  'Young' = setdiff(young_switch_genes, old_switch_genes),
  'Old' = setdiff(old_switch_genes, young_switch_genes),
  'Shared' = intersect(young_switch_genes, old_switch_genes)
)


# melt matrix for plotting:ff
tmp1 <- young_ordered[rownames(young_ordered) %in% input_list$Young,]
plot_df <- reshape2::melt(as.matrix( tmp1[rev(rownames(tmp1)),]))


p1 <- plot_df %>% 
  ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='seagreen') +
  scale_fill_gradient2(low='white', high='seagreen') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 



# melt matrix for plotting:ff
tmp2 <- old_ordered[rownames(old_ordered) %in% input_list$Old,]
plot_df <- reshape2::melt(as.matrix( tmp2[rev(rownames(tmp2)),]))


p2 <- plot_df %>% 
  ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='darkorchid3') +
  scale_fill_gradient2(low='white', high='darkorchid3') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 



# melt matrix for plotting:ff
tmp1 <- young_ordered[rownames(young_ordered) %in% input_list$Shared,]
plot_df <- reshape2::melt(as.matrix( tmp1[rev(rownames(tmp1)),]))


p3 <- plot_df %>% 
  ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  ggrastr::rasterise(geom_tile(), dpi=300) +
  scale_color_gradient2(low='white', high='black') +
  scale_fill_gradient2(low='white', high='black') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_text(angle=0, face='bold', vjust=0.5),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  ) + ylab('') 




patch <- (p1 + p3 + p2) + plot_layout(guides='collect')

pdf(paste0(fig_dir, cur_celltype, '_switchgene_heatmap.pdf'), width=6, height=6)
print(patch)
dev.off()


#


#----------------------------------------------------------------#
# Gene Overlap
#----------------------------------------------------------------#

library(eulerr)
library(GeneOverlap)

overlap_list <- c(
  'Old' = length(input_list$Old),
  'Young' =  length(input_list$Young),
  'Old&Young' =  length(input_list$Shared)
)


pdf(paste0(fig_dir, cur_celltype, '_switch_age_overlaps.pdf'), width=4, height=4)
plot(euler(overlap_list), quantities = TRUE)
dev.off()



cur_overlap <- testGeneOverlap(newGeneOverlap(
    young_switch_genes,
    old_switch_genes,
    genome.size=nrow(seurat_rna)
))

print(cur_overlap)

```

Correlation of pseudotime ranks between Young and Aged 

```{r eval=FALSE}


#library(ggnetwork)
#library(network)
#library(sna)
library(igraph)
library(tidygraph)
library(ggraph)
#library(intergraph)
library(hdWGCNA)

validation_tfs <- c('Sox10', 'Nkx2-2', 'Nkx6-2', 'Tcf4', 'Olig1', 'Olig2', 'Bach2', 'Sox8', 'Elf2', 'Bhlhe41', 'Nr6a1', 'Stat3', 'Foxk2')


# define the grey genes
modules_oligo <- GetModules(seurat_hd, wgcna_name = 'oligo')
modules_opc <- GetModules(seurat_hd, wgcna_name = 'opc')
grey_genes <- unique(c(
  subset(modules_oligo, module == 'grey') %>% .$gene_name,
  subset(modules_opc, module == 'grey') %>% .$gene_name
))
grey_genes <- grey_genes[!(grey_genes %in% c(validation_tfs))]


seurat_rna <- SetupForWGCNA(
  seurat_rna,
  gene_select = 'fraction',
  fraction = 0.05,
  group.by = 'pseudotime_bins_50',
  wgcna_name = "test"
)

expressed_genes <- GetWGCNAGenes(seurat_rna)
length(expressed_genes)

# define the number of pseudotime bins
nbins <- 50

pathfinding_tfs <- list()
cur_group <- 'Young'; plot_colors <- c('seagreen', 'seagreen3')

load(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_expression_trajectory_data.rda'))
importance_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_xgboost_importance.csv') )
target_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_tf_target_df.csv') ) %>% dplyr::select(-X)

tfs <- as.character(tf_z_df$gene)
target_df <- subset(target_df, source %in% tfs)
full_tf_z_df <- tf_z_df

# normalized rank:
full_tf_z_df$rank_norm <- full_tf_z_df$rank / nrow(z_df)

young_tf_z_df <- subset(full_tf_z_df, (gene %in% tfs) & !(gene %in% grey_genes))


cur_group <- 'Old'; plot_colors <- c('darkorchid3', 'mediumorchid1')


load(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_expression_trajectory_data.rda'))
importance_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_xgboost_importance.csv') )
target_df <- read.csv(file=paste0("/home/groups/singlecell/smorabito/PhD_ongoing/AMRF/processed/", cur_celltype, '_', cur_group, '_tf_target_df.csv') ) %>% dplyr::select(-X)

tfs <- as.character(tf_z_df$gene)
target_df <- subset(target_df, source %in% tfs)
full_tf_z_df <- tf_z_df

# normalized rank:
full_tf_z_df$rank_norm <- full_tf_z_df$rank / nrow(z_df)

old_tf_z_df <- subset(full_tf_z_df, (gene %in% tfs) & !(gene %in% grey_genes))

#------------------------------------------------------#
# Compare ranks
#------------------------------------------------------#

genes_keep <- intersect(young_tf_z_df$gene, old_tf_z_df$gene)

young_tf_z_df <- young_tf_z_df[genes_keep,]
old_tf_z_df <- old_tf_z_df[genes_keep,]

plot_df <- data.frame(
  gene = young_tf_z_df$gene,
  young_rank = young_tf_z_df$rank_norm, 
  old_rank = old_tf_z_df$rank_norm
)

min_val <- min(min(plot_df$young_rank), min(plot_df$old_rank))

plot_df$FC <- ((plot_df$young_rank + min_val) - (plot_df$old_rank + min_val)) / plot_df$young_rank
quantile(plot_df$FC)

plot_df$diff <- plot_df$young_rank - plot_df$old_rank

# calcualte the correlation of the young vs. old ranks!
cor.test(plot_df$young_rank, plot_df$old_rank)
# 0.5576409
# p-value < 2.2e-16

# how many instances where old > young and vice versa 
old_greater <- sum(plot_df$old_rank > plot_df$young_rank)

# which to label?
plot_df$lab <- ifelse(plot_df$gene %in% validation_tfs, as.character(plot_df$gene), "")

# set up plot annotations:



  up_left <- sum(plot_df$old_rank > plot_df$young_rank)
  down_right <- sum(plot_df$old_rank < plot_df$young_rank)

  annotations <- data.frame(
        xpos = c(-Inf,Inf),
        ypos =  c(Inf,-Inf),
        annotateText = c(as.character(up_left), as.character(down_right)),
        hjustvar = c(-1,2),
        vjustvar = c(2,-1)) #<- adjust



p <- plot_df %>% 
  ggplot(aes(x = young_rank, y = old_rank, fill=diff)) + 
  geom_abline() +
  ggrastr::rasterise(geom_point(
    data = plot_df %>% subset(lab == ""),
    aes(color=diff), size=3
    ), dpi=500) + 
  geom_point(
    data = plot_df %>% subset(lab != ""),
    shape = 21, color='black', size=4
  ) + 
  ggrepel::geom_text_repel(aes(label=lab), max.overlaps=Inf, fill="white", fontface='italic') +
  geom_text(inherit.aes=FALSE, data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText)) +
  scale_fill_gradient2(high='seagreen3', low='darkorchid3', mid='lightgrey') + 
  scale_color_gradient2(high='seagreen3', low='darkorchid3', mid='lightgrey') + 
  coord_fixed() + 
  theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(fill=NA, color='black')
  ) + 
  ylab("Trajectory rank, Aged") + xlab("Trajectory rank, Young")


pdf(paste0(fig_dir, cur_celltype, '_tf_ranks_compare.pdf'), width=6, height=4)
print(p)
dev.off()

```
