
```{r eval=FALSE}

# conda activate cicero
library(Seurat)
library(Signac)
library(tidyverse)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(ggrepel)
library(ggpubr)

set.seed(12345)
setwd("/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC")

data_dir <- 'data/'
fig_dir <- 'figures/'

umap_theme <- theme(
  axis.line=element_blank(),
  axis.text.x=element_blank(),
  axis.text.y=element_blank(),
  axis.ticks=element_blank(),
  axis.title.x=element_blank(),
  axis.title.y=element_blank(),
  panel.background=element_blank(),
  panel.border=element_blank(),
  panel.grid.major=element_blank(),
  panel.grid.minor=element_blank()
)

# set the current cell-type (oligos)
cur_celltype <- 'ODC'

################################################################################
# load Seurat objects
################################################################################

# load the co-embedding object for the oligo lineage:
seurat_obj <- readRDS('/dfs7/swaruplab/smorabit/collab/AMRF/analysis//ATAC/data/ODC_coembed_seurat.rds')

# load the ATAC & RNA objects separately, and subset just by the oligo lineage cells:
seurat_atac <- readRDS('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/signac_final_07-02-21.rds')
DefaultAssay(seurat_atac) <- 'peaks'
seurat_atac$Time.point <- seurat_atac$Timepoint

seurat_rna <- readRDS("/dfs7/swaruplab/smorabit/collab/AMRF/analysis/snRNAseq/NucSeq_final_doublets_removed_online_inmf_seurat.rds")

atac_ix <- na.omit(match(colnames(seurat_obj), colnames(seurat_atac)))
rna_ix <- na.omit(match(colnames(seurat_obj), colnames(seurat_rna)))

# subset by atac/rna cells that are in the oligo pseudotime object:
seurat_atac <- seurat_atac[,atac_ix]
seurat_rna <- seurat_rna[,rna_ix]

# add the co-embed umap to atac/rna objects:
atac_ix <- na.omit(match(colnames(seurat_atac), colnames(seurat_obj)))
rna_ix <- na.omit(match(colnames(seurat_rna), colnames(seurat_obj)))
atac_umap <- seurat_obj@reductions$umap@cell.embeddings[atac_ix,]
rna_umap <- seurat_obj@reductions$umap@cell.embeddings[rna_ix,]
seurat_rna@meta.data <- cbind(seurat_rna@meta.data, seurat_obj@meta.data[rna_ix,c('pseudotime', 'monocle_clusters', 'monocle_partitions', 'pseudotime_bins_10', 'pseudotime_bins_50', 'pseudotime_bins_100')])
seurat_atac@meta.data <- cbind(seurat_atac@meta.data, seurat_obj@meta.data[atac_ix,c('pseudotime', 'monocle_clusters', 'monocle_partitions', 'pseudotime_bins_10', 'pseudotime_bins_50', 'pseudotime_bins_100')])


seurat_atac@reductions$seurat_obj_umap <- CreateDimReducObject(
  embeddings = atac_umap,
  key='UMAP_',
  assay='RNA'
)

seurat_rna@reductions$seurat_obj_umap <- CreateDimReducObject(
  embeddings = rna_umap,
  key='UMAP_',
  assay='RNA'
)

seurat_obj$subcluster_name <- factor(
  as.character(seurat_obj$subcluster_name),
  levels = c('OPC', 'NFOL', 'MF-ODC', 'Int-ODC', 'Mat-ODC')
)



################################################################################
# load differential expression / accessibility data
################################################################################

# re-load t-DEGs for this celltype:
tdegs <- read.csv(paste0(data_dir, cur_celltype, '_tDEGs.csv'))
pseudotime_genes <- tdegs$gene_id

# re-load t-DARs for this celltype:
t_dars <- read.csv(paste0(data_dir, cur_celltype, '_tDARs.csv'))

# load snRNA-seq cluster marker genes:
deg_file <- "/dfs7/swaruplab/smorabit/collab/AMRF/analysis/DEGs/data/cluster_markers.csv"
markers <- read.csv(deg_file, stringsAsFactors=FALSE)
markers <- markers %>% subset(group %in% c('NFOL', 'OPC', 'MF-ODC', 'Int-ODC', 'Mat-ODC'))
oligo_markers <- markers %>% .$gene %>% unique


################################################################################
# load co-accessibility data
################################################################################

# re-load gl-cCRE table:
link_df <- read.csv(file=paste0(data_dir, cur_celltype, '_gl-cCREs.csv'))
link_df_full <- read.csv(file=paste0(data_dir, cur_celltype, '_full_link_table.csv'))
peak1_ranges <- Signac::StringToGRanges(link_df$Peak1, sep=c('-', '-'))
peak1_ranges$peak <- link_df$Peak1

peak2_ranges <- Signac::StringToGRanges(link_df$Peak2, sep=c('-', '-'))
peak2_ranges$peak <- link_df$Peak2

################################################################################
# load GVIZ / cicero plotting stuff
################################################################################

library(EnsDb.Mmusculus.v79)
library(Gviz)
library(cicero)

# load gene annotation for mm10
gene.coords <- genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "protein_coding")
genebody.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')
genebodyandpromoter.coords <- Extend(x = genebody.coords, upstream = 2000, downstream = 0)
genebodyandpromoter.coords <- genebodyandpromoter.coords %>% subset(seqnames %in% c(1:20,'Y','X'))

# load .gtf
gtf_file = "/dfs7/swaruplab/smorabit/resources/cellranger-atac_reference/refdata-cellranger-atac-mm10-1.2.0/genes/genes.gtf"
gene_anno <- rtracklayer::readGFF(gtf_file)

# rename some columns to match requirements
gene_anno$chromosome <- gene_anno$seqid
gene_anno$gene <- gene_anno$gene_id
gene_anno$transcript <- gene_anno$transcript_id
gene_anno$symbol <- gene_anno$gene_name

# testing
options(ucscChromosomeNames=FALSE)
grtrack <- GeneRegionTrack(
  gene.coords, genome = 'mm10', chromosome = 'chr10',
  name = "Gene Model",
  transcriptAnnotation = "symbol",
  background.title = "brown", fill = 'lightgray'
)

################################################################################
# change directory for current analysis
################################################################################

setwd("/dfs7/swaruplab/smorabit/collab/AMRF/analysis/bioprocesses/")

```


subsetting:

```{r eval=FALSE}

# subset old/young:
cur_group <- 'Young'
cur_group <- 'Old'

subset.cells <- ((seurat_obj$Time.point == 'Naive' & seurat_obj$subcluster_name == 'OPC') |
(seurat_obj$Time.point == 'Day.5' & seurat_obj$subcluster_name %in% c('OPC', 'NFOL')) |
(seurat_obj$Time.point == 'Day.14' & seurat_obj$subcluster_name == 'Int-ODC') |
(seurat_obj$Time.point == 'Day.30' & seurat_obj$subcluster_name == 'Mat-ODC')) &
(seurat_obj$Group == cur_group)
table(subset.cells)

# subset the seurat obj
seurat_subset <- seurat_obj[,subset.cells]

```

Genes & TFs of interest:

```{r eval=FALSE}

# TF list from collaborators:
tfs <- list(
  'cholesterol' = c(
    'Srebf1','Srebf2','Ppara',
    'Ppard','Pparg','Rxr', 'Nr1h3','Nr1h2','Rxra','Rxrb','Rxrg','Xbp'
  ),
  'oligo' = c(
    'Sox10', 'Olig1','ZFP536','Olig2','Myrf',
    'Nkx2-2','Yy1','ZNF24','Smad7','Nkx6-2','Id4',  'Id2','TCF4'
  )
)

# gene list from collaborators:
genes <- list(
  'MF-ODC_14dpi' = c(
    # TFs
    'Olig2','Zfp24','Tcf4','Srebf1','Sox10',
    # OL markers
    'Gstp1','Dcc','Tmem163',
    # myelination
    'Cnp','Plp1','Mog','Omg',
    # cholesterol biosynthesis
    'Hmgcs1','Fdft1','Insig1','Idi1','Sqle','Fdps','Cyp51','Lss','Msmo1','Sc5d',
    # lipid biosynthesis
    'Acsl1',  'Fasn','Scd2','Elovl1','Sc5d'
  ),
  'Int-ODC_14dpi' = c(
    # TFs
    'Olig1','Zfp536','Sox10',
    # ODC markers
    'Enpp6','Tmem163','Dcc','Gstp1',
    # myelination
    'Cnp','Plp1','Mobp','Mog','Opalin','Mbp',
    # cholesterol biosynthesis
    'Hmgcs1','Fdft1','Fdps','Idi1','Sqle','Dhcr24','Cyp51','Msmo1','Sc5d',
    # lipid biosynthesis
    'Scd2','Acsl3','Degs1'
  ),
  'Mat-ODC_30dpi' = c(
    #TFs
    'Sox10','Zfp536',
    # ODC markers
    'Dcc',
    # Myelination markers
    'Mobp','Mog','Mag','Plp1',
    # cholesterol biosynthesis
    'Fdft1','Dhcr24','Nsdhl','Lss','Msmo1',
    # lipid biosynthesis
    'Acaca','Elovl7','Scd2','Cpt2'
  )
)


```


Gene Module scores for specific biological processes:

```{r eval=FALSE}

# load the bioprocesses that we downloaded from the enrichr site:
enrichr_files <- dir('data/bioprocesses')

bioprocess_list <- list()
for(cur_file in enrichr_files){

  temp <- read.table(paste0('data/bioprocesses/', cur_file), sep='\t') %>% as.character %>% na.omit
  temp <- temp[temp != "NA"]

  print(cur_file)
  print(head(temp))

  # which index is the Term ID?
  term_ix <- which(grepl('[(]', temp))

  if(length(term_ix) == 0){
    print('going here')
    term_ix <- which(grepl('[WP]', temp))[1]
  }

  # split between the Term name & the genes
  print(term_ix)
  term <- paste(temp[1:term_ix], collapse='_')
  cur_genes <- temp[(term_ix+1):length(temp)]

  # convert to lowercase because mouse
  cur_genes <-  str_to_title(cur_genes)

  # check that the genes are in the Seurat object:
  cur_genes <- cur_genes[cur_genes %in% rownames(seurat_obj)]
  bioprocess_list[[term]] <- cur_genes
}

################################################################################
# compute module scores:
################################################################################

library(UCell)

# compute module scores
gene_scores <- UCell::AddModuleScore_UCell(
  seurat_obj,
  features = bioprocess_list
)@meta.data

# re-format dataframe
gene_scores <- gene_scores[,!(colnames(gene_scores) %in% colnames(seurat_obj@meta.data))]
colnames(gene_scores) <- names(bioprocess_list)


################################################################################
# bioprocess trajectories
################################################################################

plot_df <- seurat_obj@meta.data[,c('Group', 'Time.point', 'pseudotime')]
plot_df$value <- gene_scores[,cur_process]

p <- plot_df %>% ggplot(aes(x = pseudotime, y=value)) +
  geom_point(pt.size=0.2, alpha=0.2) +
  geom_smooth()

pdf(paste0(fig_dir, 'test_bioprocess_pseudotime.pdf'), width=7, height=3)
p
dev.off()


for(cur_process in names(gene_scores)){

  print(cur_process)

  plot_df <- seurat_obj@meta.data[,c('Group', 'Time.point', 'pseudotime', 'pseudotime_bins_50')]
  plot_df$value <- gene_scores[,cur_process]
  plot_df$Group.Time <- paste0(as.character(plot_df$Group), '_', as.character(plot_df$Time.point))

  # compute the average score for each bin & each group:
  groups <- unique(plot_df$Group.Time)
  bins <- levels(plot_df$pseudotime_bins_50)
  avg_scores <- sapply(groups, function(x){
    sapply(bins, function(y){
      mean(plot_df[plot_df$Group.Time == x & plot_df$pseudotime_bins_50 == y,'value'])
    })
  })
  colnames(avg_scores) <- groups; rownames(avg_scores) <- bins

  # replace NAs with 0:
  avg_scores[is.na(avg_scores)] <- 0

  # convert to long format for plotting
  avg_df <- reshape2::melt(avg_scores) %>% dplyr::rename(c(pseudotime_bins_50=Var1, Group.Time=Var2))
  avg_df$pseudotime_bins_50 <- factor(
    as.character(avg_df$pseudotime_bins_50),
    levels = levels(seurat_obj$pseudotime_bins_50)
  )
  avg_df$Group.Time <- factor(
    as.character(avg_df$Group.Time),
    levels = c(
      'Young_Naive', 'Young_Day.5', 'Young_Day.14', 'Young_Day.30',
      'Old_Naive', 'Old_Day.5', 'Old_Day.14', 'Old_Day.30'
    )
  )

  # add a column to for values scaled between 0 & 1:
  avg_df <- avg_df %>% group_by(Group.Time) %>% mutate(scaled = scale01(value))

  p <- avg_df %>% ggplot(aes(x = as.numeric(pseudotime_bins_50), y=value)) +
    geom_smooth() +
    geom_point(size=0.2) +
    theme(
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) + xlab('pseudotime') + ylab('UCell score') +
    ggtitle(cur_process)


  pdf(paste0(fig_dir, 'bioprocess_trajectories/', gsub(' ', '_', cur_process), '_pseudotime.pdf'), width=8, height=4)
  print(p + facet_wrap(~Group.Time, ncol=4))
  dev.off()


  # plot scaled
  p <- avg_df %>% ggplot(aes(x = as.numeric(pseudotime_bins_50), y=scaled)) +
    geom_smooth() +
    geom_point(size=0.2) +
    theme(
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) + xlab('pseudotime') + ylab('UCell score') +
    ggtitle(cur_process)


  pdf(paste0(fig_dir, 'bioprocess_trajectories/', gsub(' ', '_', cur_process), '_pseudotime_scaled.pdf'), width=8, height=4)
  print(p + facet_wrap(~Group.Time, ncol=4))
  dev.off()


}


```

Get linked enhancers for these bioprocesses:

```{r eval=FALSE}

scale01 <- function(x){
  y <- min(x); z <- max(x);
  (x-y) / (z-y)
}


################################################################################
# Get enhancers linked to genes in each biological process:
################################################################################

n_enhancers <- 150
terms <- names(bioprocess_list)
bioprocess_link_df <- data.frame()
for(cur_term in terms){

  cur_genes <- bioprocess_list[[cur_term]]

  # only get genes in our link df
  cur_genes <- cur_genes[cur_genes %in% link_df$Peak1_nearestGene]

  # loop through genes and get linked cCREs:
  cur_link_df <- data.frame()
  for(cur_gene in cur_genes){

    #print(cur_gene)

    # get the cCREs linked to this gene:
    temp <- link_df %>% subset(Peak1_nearestGene == cur_gene & Peak2_type == 'Distal')
    print(dim(temp))

    # add these cCREs to our list
    cur_link_df <- rbind(cur_link_df, temp)

  }
  cur_link_df$term <- cur_term

  cur_link_df <- cur_link_df %>% top_n(n_enhancers, wt=coaccess)

  bioprocess_link_df <- rbind(bioprocess_link_df, cur_link_df)
}
table(bioprocess_link_df$term)

################################################################################
# Compute Module Scores:
################################################################################

# format for module score:
enhancer_list <- list()
for(cur_term in terms){
  enhancer_list[[cur_term]] <- bioprocess_link_df %>% subset(term == cur_term) %>% .$Peak2 %>% unique
}

# # compute module scores
# enhancer_scores <- UCell::AddModuleScore_UCell(
#   seurat_atac,
#   features = enhancer_list
# )@meta.data

# compute module scores
enhancer_scores <- UCell::AddModuleScore_UCell(
  seurat_atac,
  features = enhancer_list
)@meta.data

# re-format dataframe
enhancer_scores <- enhancer_scores[,!(colnames(enhancer_scores) %in% colnames(seurat_atac@meta.data))]
colnames(enhancer_scores) <- terms


################################################################################
# bioprocess trajectories enhancers
################################################################################

for(cur_process in names(enhancer_scores)){

  print(cur_process)

  plot_df <- seurat_atac@meta.data[,c('Group', 'Time.point', 'pseudotime', 'pseudotime_bins_50')]
  plot_df$value <- enhancer_scores[,cur_process]
  plot_df$Group.Time <- paste0(as.character(plot_df$Group), '_', as.character(plot_df$Time.point))

  # compute the average score for each bin & each group:
  groups <- unique(plot_df$Group.Time)
  bins <- levels(plot_df$pseudotime_bins_50)
  avg_scores <- sapply(groups, function(x){
    sapply(bins, function(y){
      mean(plot_df[plot_df$Group.Time == x & plot_df$pseudotime_bins_50 == y,'value'])
    })
  })
  colnames(avg_scores) <- groups; rownames(avg_scores) <- bins

  # replace NAs with 0:
  avg_scores[is.na(avg_scores)] <- 0

  # convert to long format for plotting
  avg_df <- reshape2::melt(avg_scores) %>% dplyr::rename(c(pseudotime_bins_50=Var1, Group.Time=Var2))
  avg_df$pseudotime_bins_50 <- factor(
    as.character(avg_df$pseudotime_bins_50),
    levels = levels(seurat_obj$pseudotime_bins_50)
  )
  avg_df$Group.Time <- factor(
    as.character(avg_df$Group.Time),
    levels = c(
      'Young_Naive', 'Young_Day.5', 'Young_Day.14', 'Young_Day.30',
      'Old_Naive', 'Old_Day.5', 'Old_Day.14', 'Old_Day.30'
    )
  )


  # add a column to for values scaled between 0 & 1:
  avg_df <- avg_df %>% group_by(Group.Time) %>% mutate(scaled = scale01(value))

  # plot scaled trajectories
  p <- avg_df %>% ggplot(aes(x = as.numeric(pseudotime_bins_50), y=scaled)) +
    geom_smooth(color='seagreen') +
    geom_point(size=0.2) +
    theme(
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) + xlab('pseudotime') + ylab('Scaled UCell score') +
    ggtitle(cur_process)

  pdf(paste0(fig_dir, 'bioprocess_trajectories/', gsub(' ', '_', cur_process), '_enhancer_pseudotime_scaled.pdf'), width=8, height=4)
  print(p + facet_wrap(~Group.Time, ncol=4))
  dev.off()

  # plot un-scaled trajectories
  p <- avg_df %>% ggplot(aes(x = as.numeric(pseudotime_bins_50), y=value)) +
    geom_smooth(color='seagreen') +
    geom_point(size=0.2) +
    theme(
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      plot.title = element_text(hjust = 0.5)
    ) + xlab('pseudotime') + ylab('UCell score') +
    ggtitle(cur_process)

  pdf(paste0(fig_dir, 'bioprocess_trajectories/', gsub(' ', '_', cur_process), '_enhancer_pseudotime.pdf'), width=8, height=4)
  print(p + facet_wrap(~Group.Time, ncol=4))
  dev.off()

}

















################################################################################
# Visualize on the UMAP:
################################################################################

umap_df <- seurat_atac@reductions$seurat_obj_umap@cell.embeddings %>% as.data.frame()

cur_term <- terms[4]

plot_df <- umap_df
plot_df$score <- as.numeric(enhancer_scores[,cur_term])


plot_df <- arrange(plot_df, score)

p <- plot_df %>%
    ggplot(aes(x=UMAP_1, y=UMAP_2, color=score)) +
    geom_point(size=1, alpha=0.75) +
    ggtitle(cur_term) + umap_theme +
    labs(color="")+ scale_color_gradient2(
        low='grey75', mid='grey95', high='purple',
      #  breaks = plot_range,
        #labels = c('-', '+'),
        guide = guide_colorbar(ticks=FALSE, barwidth=0.5, barheight=4)
      )


pdf(paste0(fig_dir, 'test_enhancer_score.pdf'), width=7, height=7)
p
dev.off()


################################################################################
# what happens if I use Rmagic??
################################################################################

library(Rmagic)
data(magic_testdata)
MAGIC_data <- Rmagic::magic(magic_testdata, genes=c("VIM", "CDH1", "ZEB1"))

test <- Rmagic::magic(enhancer_scores, genes=colnames(enhancer_scores))


umap_df <- seurat_atac@reductions$seurat_obj_umap@cell.embeddings %>% as.data.frame()

cur_term <- terms[4]

plot_df <- umap_df
plot_df$score <- as.numeric(test$result[,cur_term])


plot_df <- arrange(plot_df, score)

p <- plot_df %>%
    ggplot(aes(x=UMAP_1, y=UMAP_2, color=score)) +
    geom_point(size=1, alpha=0.75) +
    ggtitle(cur_term) + umap_theme +
    labs(color="")+ scale_color_gradient2(
        low='grey75', mid='grey95', high='purple',
      #  breaks = plot_range,
        #labels = c('-', '+'),
        guide = guide_colorbar(ticks=FALSE, barwidth=0.5, barheight=4)
      )


pdf(paste0(fig_dir, 'test_enhancer_score_magic.pdf'), width=7, height=7)
p
dev.off()


```



subsetting:

```{r eval=FALSE}

# subset old/young:
cur_group <- 'Young'
cur_group <- 'Old'

load(file='/dfs7/swaruplab/smorabit/collab/AMRF/data/color_scheme.rda')


subset.cells <- ((seurat_obj$Time.point == 'Naive' & seurat_obj$subcluster_name == 'OPC') |
(seurat_obj$Time.point == 'Day.5' & seurat_obj$subcluster_name %in% c('OPC', 'NFOL')) |
(seurat_obj$Time.point == 'Day.14' & seurat_obj$subcluster_name %in% c('MF-ODC', 'Int-ODC')) |
(seurat_obj$Time.point == 'Day.30' & seurat_obj$subcluster_name == 'Mat-ODC')) &
(seurat_obj$Group == cur_group)
table(subset.cells)

# subset the seurat obj
seurat_subset <- seurat_obj[,subset.cells]

#
# subset.cells <- ((seurat_rna$Time.point == 'Naive' & seurat_rna$subcluster_name == 'OPC') |
# (seurat_rna$Time.point == 'Day.5' & seurat_rna$subcluster_name %in% c('OPC', 'NFOL')) |
# (seurat_rna$Time.point == 'Day.14' & seurat_rna$subcluster_name %in% c('MF-ODC', 'Int-ODC')) |
# (seurat_rna$Time.point == 'Day.30' & seurat_rna$subcluster_name == 'Mat-ODC')) &
# (seurat_rna$Group == cur_group)
# table(subset.cells)
#
# # subset the seurat obj
# seurat_subset <- seurat_rna[,subset.cells]
#




# cut pseudotime into bins:
bins_10 <- seurat_subset$pseudotime %>% Hmisc::cut2(g=10)
bins_50 <- seurat_subset$pseudotime %>% Hmisc::cut2(g=50)
bins_100 <- seurat_subset$pseudotime %>% Hmisc::cut2(g=100)

seurat_subset$pseudotime_bins_10 <- bins_10
seurat_subset$pseudotime_bins_50 <- bins_50
seurat_subset$pseudotime_bins_100 <- bins_100


```


curated process:

```{r eval=FALSE}


scale01 <- function(x){
  y <- min(x); z <- max(x);
  (x-y) / (z-y)
}

################################################################################
# Older bioprocesses
################################################################################

curated_file <- "/dfs7/swaruplab/smorabit/collab/AMRF/analysis/bioprocesses/data/curated_pathways.txt"
n_lines <- 4
tmp <- readLines(curated_file, n=n_lines)


bioprocess_list <- list()
for(i in 1:n_lines){
  cur <- tmp[[i]]
  cur <- str_split(cur, '\t')[[1]]
  cur_genes <- unique(cur[-1])

  # convert to lowercase because mouse
  cur_genes <-  str_to_title(cur_genes)

  cur_genes <- cur_genes[cur_genes %in% rownames(seurat_rna)]
  bioprocess_list[[cur[1]]] <- cur_genes

}

################################################################################
# Newer bioprocesses August 2022
################################################################################

pathway_dir <- "/dfs7/swaruplab/smorabit/collab/AMRF/analysis/bioprocesses/data/gene_lists/"
pathway_files <- dir(pathway_dir)

# get pathway names
pathway_names <- gsub('.txt', '', pathway_files)
pathway_names <- gsub(' ', '_', pathway_names)

bioprocess_list <- lapply(1:length(pathway_names), function(i){
  cur_genes <- read.table(paste0(pathway_dir, pathway_files[i]), header=FALSE) %>% .$V1
  cur_genes <-  str_to_title(cur_genes)
  cur_genes <- cur_genes[cur_genes %in% rownames(seurat_rna)]
  cur_genes
})

names(bioprocess_list) <- pathway_names


################################################################################
# compute module scores:
################################################################################

library(UCell)

# compute module scores
gene_scores <- UCell::AddModuleScore_UCell(
  seurat_subset,
  features = bioprocess_list
)@meta.data

# re-format dataframe
gene_scores <- gene_scores[,(ncol(gene_scores) - length(bioprocess_list) + 1):ncol(gene_scores)]
colnames(gene_scores) <- names(bioprocess_list)


################################################################################
# bioprocess trajectories
################################################################################

library(patchwork)
library(ggstream)


# proportion by time point
proportion_df <- seurat_subset@meta.data %>%
  group_by(pseudotime_bins_50, Time.point) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n))

prop_tp <- ggplot(proportion_df, aes(x = as.numeric(pseudotime_bins_50), y = freq, fill = Time.point)) +
  geom_stream(type = "proportion") +
  scale_fill_brewer(palette='PuBu', direction=-1) +
  theme(
    axis.line=element_blank(),
    axis.text.x=element_blank(),
    axis.text.y=element_blank(),
    axis.ticks=element_blank(),
    axis.title.x=element_blank(),
    axis.title.y=element_blank(),
    panel.background=element_blank(),
    panel.border=element_blank(),
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    plot.margin = margin(0,0,0,0)
  ) +
  xlab('Pseudotime')

# proportion by time point
proportion_df <- seurat_subset@meta.data %>%
  group_by(pseudotime_bins_50, subcluster_name) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n))

prop_ct <- ggplot(proportion_df, aes(x = as.numeric(pseudotime_bins_50), y = freq, fill = subcluster_name)) +
  geom_stream(type = "proportion") +
  #scale_fill_brewer(palette='PuBu', direction=-1) +
  theme(
    axis.line=element_blank(),
    axis.text.x=element_blank(),
    axis.text.y=element_blank(),
    axis.ticks=element_blank(),
    #axis.title.x=element_blank(),
    axis.title.y=element_blank(),
    panel.background=element_blank(),
    panel.border=element_blank(),
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    plot.margin = margin(0,0,0,0)
  ) +
  xlab('Pseudotime') + scale_fill_manual(values=cluster_colors_rna[as.character(unique(seurat_subset$subcluster_name))])

for(cur_process in names(gene_scores)){

  print(cur_process)

  plot_df <- seurat_subset@meta.data[,c('Group', 'Time.point', 'pseudotime', 'pseudotime_bins_50')]
  plot_df$value <- gene_scores[,cur_process]
  plot_df$Group.Time <- paste0(as.character(plot_df$Group), '_', as.character(plot_df$Time.point))


  # compute the average score for each bin & each group:
  avg_df <- plot_df %>% group_by(pseudotime_bins_50) %>% summarise(value=mean(value))

  # add a column to for values scaled between 0 & 1:
  avg_df <- avg_df %>% mutate(scaled = scale01(value))

  p <- avg_df %>% ggplot(aes(x = as.numeric(pseudotime_bins_50), y=value)) +
    geom_smooth() +
    geom_point(size=0.2) +
    theme(
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      axis.title.x = element_blank(),
      plot.title = element_text(hjust = 0.5),
      plot.margin = margin(0,0,0,0),
      axis.line.x = element_blank()
    ) + xlab('pseudotime') + ylab('UCell score') +
    ggtitle(cur_process)

  patch <- p / prop_tp / prop_ct + plot_layout(heights=c(5,1, 1), guides='collect')

  pdf(paste0(fig_dir, 'bioprocess_trajectories/', gsub(' ', '_', cur_process), '_', cur_group, '_pseudotime.pdf'), width=6, height=3)
  print(patch)
  dev.off()

  #
  # # plot scaled
  # p <- avg_df %>% ggplot(aes(x = as.numeric(pseudotime_bins_50), y=scaled)) +
  #   geom_smooth() +
  #   geom_point(size=0.2) +
  #   theme(
  #     axis.ticks.x = element_blank(),
  #     axis.text.x = element_blank(),
  #     plot.title = element_text(hjust = 0.5)
  #   ) + xlab('pseudotime') + ylab('UCell score') +
  #   ggtitle(cur_process)
  #
  # patch <- p / prop_tp / prop_ct + plot_layout(heights=c(5,1, 1), guides='collect')
  #
  # pdf(paste0(fig_dir, 'bioprocess_trajectories/', gsub(' ', '_', cur_process), '_', cur_group, '_pseudotime_scaled.pdf'), width=6, height=3)
  # print(patch)
  # dev.off()

}

```


Run same block as above but with Young & Old together:

```{r eval=FALSE}

#s subset

subset.cells <- ((seurat_obj$Time.point == 'Naive' & seurat_obj$subcluster_name == 'OPC') |
(seurat_obj$Time.point == 'Day.5' & seurat_obj$subcluster_name %in% c('OPC', 'NFOL')) |
(seurat_obj$Time.point == 'Day.14' & seurat_obj$subcluster_name %in% c('MF-ODC', 'Int-ODC')) |
(seurat_obj$Time.point == 'Day.30' & seurat_obj$subcluster_name == 'Mat-ODC'))
table(subset.cells)

# subset the seurat obj
seurat_subset <- seurat_obj[,subset.cells]


# cut pseudotime into bins:
bins_10 <- seurat_subset$pseudotime %>% Hmisc::cut2(g=10)
bins_50 <- seurat_subset$pseudotime %>% Hmisc::cut2(g=50)
bins_100 <- seurat_subset$pseudotime %>% Hmisc::cut2(g=100)

seurat_subset$pseudotime_bins_10 <- bins_10
seurat_subset$pseudotime_bins_50 <- bins_50
seurat_subset$pseudotime_bins_100 <- bins_100

# now subset by young and old
seurat_y <- subset(seurat_subset, Group == 'Young')
seurat_o <- subset(seurat_subset, Group == 'Old')

################################################################################
# Newer bioprocesses August 2022
################################################################################

pathway_dir <- "/dfs7/swaruplab/smorabit/collab/AMRF/analysis/bioprocesses/data/gene_lists/"
pathway_files <- dir(pathway_dir)

# get pathway names
pathway_names <- gsub('.txt', '', pathway_files)
pathway_names <- gsub(' ', '_', pathway_names)

bioprocess_list <- lapply(1:length(pathway_names), function(i){
  cur_genes <- read.table(paste0(pathway_dir, pathway_files[i]), header=FALSE) %>% .$V1
  cur_genes <-  str_to_title(cur_genes)
  cur_genes <- cur_genes[cur_genes %in% rownames(seurat_rna)]
  cur_genes
})

names(bioprocess_list) <- pathway_names


################################################################################
# compute pathway scores
################################################################################

library(UCell)

# compute module scores
gene_scores_y <- UCell::AddModuleScore_UCell(seurat_y, features = bioprocess_list)@meta.data
gene_scores_o <- UCell::AddModuleScore_UCell(seurat_o, features = bioprocess_list)@meta.data

# re-format dataframe
gene_scores_y  <- gene_scores_y [,(ncol(gene_scores_y ) - length(bioprocess_list) + 1):ncol(gene_scores_y )]
gene_scores_o  <- gene_scores_o [,(ncol(gene_scores_o ) - length(bioprocess_list) + 1):ncol(gene_scores_o )]
colnames(gene_scores_y ) <- names(bioprocess_list)
colnames(gene_scores_o ) <- names(bioprocess_list)


# combine young & old
gene_scores <- rbind(gene_scores_y, gene_scores_o)
gene_scores <- gene_scores[colnames(seurat_subset),]

all.equal(rownames(gene_scores), colnames(seurat_subset))

# add gene scores as a new assay:
gene_score_assay <- Seurat::CreateAssayObject(t(gene_scores))


################################################################################
# compute  differential pathway scores by group
################################################################################



################################################################################
# compute  differential pathway scores by group
################################################################################


# groups <- levels(seurat_subset$pseudotime_bins_10)
groups <- levels(seurat_subset$subcluster_name)
dp_list <- list()
for(cur_group in groups){


  # group1 (young)
  barcodes1 <- seurat_subset@meta.data %>% subset(Group == 'Young' & subcluster_name == cur_group) %>% rownames
  barcodes2 <- seurat_subset@meta.data %>% subset(Group == 'Old' & subcluster_name == cur_group) %>% rownames

  # run differential test on the ME assay
  cur_dp <- Seurat::FindMarkers(
    gene_score_assay,
    cells.1 = barcodes1,
    cells.2 = barcodes2,
    slot='counts',
    test.use='wilcox',
    only.pos=FALSE,
    logfc.threshold=0,
    min.pct=0,
    verbose=FALSE
  )
  cur_dp$group <- cur_group
  cur_dp$pathway <- rownames(cur_dp)

  print(range(cur_dp$avg_log2FC))

  # fix infs:
  cur_dp$avg_log2FC <- ifelse(abs(cur_dp$avg_log2FC) == Inf, 0, cur_dp$avg_log2FC)
  dp_list[[cur_group]] <- cur_dp


}
diff_pathways <- do.call(rbind, dp_list)

diff_pathways$group <- factor(
  as.character(diff_pathways$group),
  levels=groups
)

# point shape
diff_pathways$shape <- ifelse(diff_pathways$p_val_adj < 0.05, 21, 4)

# plot the diff_pathways:
p <- diff_pathways %>%
  ggplot(aes(x = group, y = avg_log2FC, fill=group, color=group)) +
  geom_hline(yintercept=0, linetype='dashed', color='darkgrey') +
  geom_segment(aes(y=0, yend=avg_log2FC, x=group, xend=group), size=0.5, color='grey') +
  geom_point(size=3) +
  geom_point(shape=diff_pathways$shape, color='black', fill=NA, size=3) +
  scale_y_continuous(limits=c(-1, 2.5)) +
  scale_color_manual(values=plasma(length(groups)), guide='none') +
  ylab(bquote("Average log"[2]~"(Fold Change)")) +
  theme(
    axis.text.x = element_blank(),
    axis.title.x = element_blank(),
    axis.ticks.x = element_blank()
  ) + NoLegend()


pdf(paste0(fig_dir, 'test_diff_pathways.pdf'), width=6, height=12)
p + facet_wrap(~pathway, ncol=1)
dev.off()




```
