
```{r eval=FALSE}

# conda activate cicero
library(Seurat)
library(Signac)
library(tidyverse)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(monocle3)
library(SeuratWrappers)
library(scales)
library(ArchR)
library(shades)

theme_set(theme_cowplot())

setwd("/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC")
data_dir <- 'data/'
fig_dir <- 'figures/'

umap_theme <- theme(
  axis.line=element_blank(),
  axis.text.x=element_blank(),
  axis.text.y=element_blank(),
  axis.ticks=element_blank(),
  axis.title.x=element_blank(),
  axis.title.y=element_blank(),
  panel.background=element_blank(),
  panel.border=element_blank(),
  panel.grid.major=element_blank(),
  panel.grid.minor=element_blank()
)

# load snATAC-seq seurat object:
seurat_obj <- readRDS(paste0(data_dir, 'signac_final_07-02-21.rds'))
DefaultAssay(seurat_obj) <- 'peaks'

# re-load data for current cell type:
cur_celltype <- 'ODC'
# cur_celltype <- 'MG'
# cur_celltype <- 'ASC'

fig_dir <- paste0('figures/', cur_celltype, '/'); dir.create(fig_dir)

seurat_subset <- readRDS(paste0(data_dir, cur_celltype, '_coembed_seurat.rds'))
seurat_celltype <- readRDS(file=paste0(data_dir, cur_celltype, '_atac_seurat.rds'))

# re-load t-DEGs for this celltype:
tdegs <- read.csv(paste0(data_dir, cur_celltype, '_tDEGs.csv'))
pseudotime_genes <- tdegs$gene_id

# re-load t-DARs for this celltype:
t_dars <- read.csv(paste0(data_dir, cur_celltype, '_tDARs.csv'))

# re-load gl-cCRE table:
link_df <- read.csv(file=paste0(data_dir, cur_celltype, '_gl-cCREs.csv'))
link_df_full <- read.csv(file=paste0(data_dir, cur_celltype, '_full_link_table.csv'))
peak1_ranges <- Signac::StringToGRanges(link_df$Peak1, sep=c('-', '-'))
peak1_ranges$peak <- link_df$Peak1

peak2_ranges <- Signac::StringToGRanges(link_df$Peak2, sep=c('-', '-'))
peak2_ranges$peak <- link_df$Peak2

# set archR genome
addArchRGenome("mm10")
addArchRThreads(threads = 8)

# re-load
proj <- loadArchRProject(path = "Franklin_ATAC")
proj@peakSet$site_name <- paste0(
  as.character(seqnames(proj@peakSet)), '-', start(proj@peakSet), '-', end(proj@peakSet)
)
proj@peakSet$peak <- paste0(
  as.character(seqnames(proj@peakSet)), '-',
  as.character(start(proj@peakSet)), '-',
  as.character(end(proj@peakSet))
)
peakset <- data.frame(
  'peak' = proj@peakSet$peak,
  'peakType' = proj@peakSet$peakType,
  'nearestGene' = proj@peakSet$nearestGene
)
rownames(peakset) <- peakset$peak


# load snRNA-seq cluster marker genes:
deg_file <- "/dfs7/swaruplab/smorabit/collab/AMRF/analysis/DEGs/data/cluster_markers.csv"
markers <- read.csv(deg_file, stringsAsFactors=FALSE)
markers <- markers %>% subset(group %in% c('NFOL', 'OPC', 'MF-ODC', 'Int-ODC', 'Mat-ODC'))
oligo_markers <- markers %>% .$gene %>% unique


```

Load stuff for Gviz, only needed for cicero plots

```{r eval=FALSE}

library(EnsDb.Mmusculus.v79)
library(Gviz)
library(cicero)


# load gene annotation for mm10
gene.coords <- genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "protein_coding")
genebody.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')
genebodyandpromoter.coords <- Extend(x = genebody.coords, upstream = 2000, downstream = 0)
genebodyandpromoter.coords <- genebodyandpromoter.coords %>% subset(seqnames %in% c(1:20,'Y','X'))

# load .gtf
gtf_file = "/dfs7/swaruplab/smorabit/resources/cellranger-atac_reference/refdata-cellranger-atac-mm10-1.2.0/genes/genes.gtf"
gene_anno <- rtracklayer::readGFF(gtf_file)

# rename some columns to match requirements
gene_anno$chromosome <- gene_anno$seqid
gene_anno$gene <- gene_anno$gene_id
gene_anno$transcript <- gene_anno$transcript_id
gene_anno$symbol <- gene_anno$gene_name

grtrack <- GeneRegionTrack(
  gene.coords, genome = 'mm10', chromosome = 'chr10',
  name = "Gene Model",
  transcriptAnnotation = "symbol",
  background.title = "brown", fill = 'lightgray'
)

```

Proportion of injury timepoints:

```{r eval=FALSE}

library(ggpubr)
theme_set(theme_pubr())

prop_df <- data.frame()

valid_pseudotime <- seurat_subset@meta.data %>% subset(pseudotime != Inf)

# Young
valid_pseudotime_age <- seurat_subset@meta.data %>% subset(pseudotime != Inf & Group == "Young")

# get proportion of cells in each bin for each time point
bins <- levels(valid_pseudotime_age$pseudotime_bins_100)
proportion_df <- data.frame()
for(i in 1:length(bins)){
  bin = bins[i]

  cur_meta <- subset(valid_pseudotime_age, pseudotime_bins_100 == bin)

  cur_df <- data.frame(table(cur_meta$Time.point) / nrow(cur_meta))
  cur_df <- cur_df %>% dplyr::rename(c(Diagnosis=Var1, proportion=Freq))
  cur_df$bin_num <- i
  cur_df$bin <- bin
  proportion_df <- rbind(proportion_df, cur_df)

}

# set factor levels:
proportion_df$Diagnosis <- factor(
  as.character(proportion_df$Diagnosis),
  levels = c('Naive', 'Day.5', 'Day.14', 'Day.30')
)
proportion_df$Diagnosis <- fct_rev(proportion_df$Diagnosis)
proportion_df$group <- 'Young'
prop_df <- rbind(prop_df, proportion_df)

p1 <- ggplot(proportion_df, aes(x=bin_num, y=proportion, fill=Diagnosis)) +
  xlab('Trajectory') + ylab('') +
  geom_bar(position='fill', stat='identity', alpha=1) +
  scale_fill_brewer(palette='PuBu', direction=-1) +
#  lightness(scalefac(0.7)) +
  # facet_wrap(~Diagnosis, ncol=1) + NoLegend() +
  theme(
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) + umap_theme


pdf(paste0(fig_dir, cur_celltype, '_monocle_coembed_proportions_young_100.pdf'), width=6, height=3, useDingbats=FALSE)
p1
dev.off()

library(ggstream)


p1 <- ggplot(proportion_df, aes(x=bin_num, y=proportion, fill=Diagnosis)) +
  xlab('Trajectory') + ylab('') +
  geom_stream(type='proportional') +
  scale_fill_brewer(palette='PuBu', direction=-1) +
  theme(
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) + umap_theme

pdf(paste0(fig_dir, cur_celltype, '_monocle_coembed_proportions_young_100_stream.pdf'), width=6, height=3, useDingbats=FALSE)
p1
dev.off()

# Old
valid_pseudotime_age <- seurat_subset@meta.data %>% subset(pseudotime != Inf & Group == "Old")

# get proportion of cells in each bin for each time point
bins <- levels(valid_pseudotime_age$pseudotime_bins_100)
proportion_df <- data.frame()
for(i in 1:length(bins)){
  bin = bins[i]

  cur_meta <- subset(valid_pseudotime_age, pseudotime_bins_100 == bin)

  cur_df <- data.frame(table(cur_meta$Time.point) / nrow(cur_meta))
  cur_df <- cur_df %>% dplyr::rename(c(Diagnosis=Var1, proportion=Freq))
  cur_df$bin_num <- i
  cur_df$bin <- bin
  proportion_df <- rbind(proportion_df, cur_df)

}

# set factor levels:
proportion_df$Diagnosis <- factor(
  as.character(proportion_df$Diagnosis),
  levels = c('Naive', 'Day.5', 'Day.14', 'Day.30')
)
proportion_df$Diagnosis <- fct_rev(proportion_df$Diagnosis)
proportion_df$group <- 'Old'
prop_df <- rbind(prop_df, proportion_df)

p2 <- ggplot(proportion_df, aes(x=bin_num, y=proportion, fill=Diagnosis)) +
  xlab('Trajectory') + ylab('') +
  geom_bar(position='fill', stat='identity', alpha=1) +
  lightness(scale_fill_brewer(palette='PuBu', direction=-1), scalefac(0.75)) +
#  lightness(scalefac(0.7)) +
  # facet_wrap(~Diagnosis, ncol=1) + NoLegend() +
  theme(
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) + umap_theme




pdf(paste0(fig_dir, cur_celltype, '_monocle_coembed_proportions_old_100.pdf'), width=6, height=3, useDingbats=FALSE)
p2
dev.off()

# stream graph
library(ggstream)


p2 <- ggplot(proportion_df, aes(x=bin_num, y=proportion, fill=Diagnosis)) +
  xlab('Trajectory') + ylab('') +
  geom_stream(type='proportional') +
  lightness(scale_fill_brewer(palette='PuBu', direction=-1), scalefac(0.75)) +
  theme(
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) + umap_theme

pdf(paste0(fig_dir, cur_celltype, '_monocle_coembed_proportions_old_100_stream.pdf'), width=6, height=3, useDingbats=FALSE)
p2
dev.off()


```

Plot reconstructed expression heatmap:

```{r eval=FALSE}

library(ComplexHeatmap)
library(future.apply)
library(viridis)


# RVAE directory
outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
cur_name <- paste0(cur_celltype, '_RVAE')

# load reconstructed data and z embedding
dataset = cur_name
recon_df <- read.table(paste0(outdir, 'reconstructed/',cur_name,'_recon.csv'), sep=',', row.names=1, header=TRUE)
z_df <- recon_df[,(ncol(recon_df)-1):ncol(recon_df)]
recon_df <- recon_df[,1:(ncol(recon_df)-2)]
z_df$gene <- rownames(z_df)

# scale between 0 & 1
range01 <- function(x){
  cur <- recon_df[x,]
  (cur-min(cur))/(max(cur)-min(cur))
}
scaled <- lapply(rownames(recon_df), range01)
scaled <- do.call(rbind, scaled)
rownames(scaled) <- rownames(recon_df)

# order rows by the time they reach 0.75% min expression
ordering <- future_lapply(1:nrow(scaled), function(i){
  match(names(scaled[i,])[scaled[i,] >= 0.75][1], colnames(scaled))
})
ordered <- scaled[rownames(scaled)[order(unlist(ordering))],]

# reverse ordering
ordered <- ordered[rownames(ordered),]
z_df$rank <- match(rownames(z_df), rownames(ordered))


# for oligos label important ODC lineage marker genes (from Marques et al 2016):
odc_lineage_genes <- c(
  'Pdgfra', 'Cspg4',
  'Vcan', 'Sox6', 'Nkx2-2',
  'Tcf7l2', 'Itpr2', 'Tmem2', 'Pdgfa',
  'Mal', 'Mog', 'Plp1', 'Opalin', 'Mbp',
  'Klk6', 'Apod'
)
cur_anno_genes <- odc_lineage_genes

# get lineage genes from DEG list:
n_degs <- 10
plot_genes <- markers %>%
  arrange(group) %>%
  subset(p_val_adj <= 0.05) %>%
  group_by(group) %>%
  top_n(n_degs, wt=avg_log2FC)  %>%
  .$gene %>% unique
odc_lineage_genes <- plot_genes
cur_anno_genes <- odc_lineage_genes


# label diagnosis cluster DEGs
degs_indices = na.omit(match(cur_anno_genes, rownames(ordered)))
degs_labels = cur_anno_genes[cur_anno_genes %in% rownames(ordered)]
degs_colors = rep('blue', length(degs_labels))
names(degs_colors) <- degs_labels
ha = rowAnnotation(
  foo = anno_mark(
    at = degs_indices, labels = degs_labels,
  ),
  col=list(foo=degs_colors),
  simple_anno_size = unit(1, "cm")
)


# plot heatmap
pdf(paste0(fig_dir, cur_celltype, '_recon_heatmap.pdf'), width=6, height=12)
ComplexHeatmap::Heatmap(
  as.matrix(ordered),
  show_column_names = FALSE, show_row_names=FALSE,
  col = viridis(256),
  cluster_rows=FALSE,
  cluster_columns=FALSE,
  right_annotation=ha,
  #left_annotation=color_ha,
  use_raster = TRUE
)
dev.off()

################################################################################
# Heatmap only using oligo markers:
################################################################################

ordered_full <- ordered
ordered <- ordered[rownames(ordered) %in% oligo_markers,]

# label diagnosis cluster DEGs
degs_indices = na.omit(match(cur_anno_genes, rownames(ordered)))
degs_labels = cur_anno_genes[cur_anno_genes %in% rownames(ordered)]
degs_colors = rep('blue', length(degs_labels))
names(degs_colors) <- degs_labels
ha = rowAnnotation(
  foo = anno_mark(
    at = degs_indices, labels = degs_labels,
    simple_anno_size = unit(1, "cm")
), col=list(foo=degs_colors))



pdf(paste0(fig_dir, cur_celltype, '_recon_heatmap_markers.pdf'), width=6, height=12)
ComplexHeatmap::Heatmap(
  as.matrix(ordered),
  show_column_names = FALSE, show_row_names=FALSE,
  col = viridis(256),
  cluster_rows=FALSE,
  cluster_columns=FALSE,
  right_annotation=ha,
  #left_annotation=color_ha,
  use_raster = TRUE
)
dev.off()


ordered <- ordered_full

################################################################################
# plot latent space colored by pseudotime rank
################################################################################
library(ggrepel)

theme_set(theme_cowplot())

z_df$rank <- match(rownames(z_df), rownames(ordered))

plot_df <- z_df
plot_df$anno <- ifelse(plot_df$gene %in% odc_lineage_genes, plot_df$gene, '')

plot_df$marker <- ifelse(plot_df$gene %in% oligo_markers, plot_df$gene, '')


p <- plot_df %>% subset(marker != '') %>%
  ggplot(aes(Z1, Z2, fill=rank)) +
  geom_point(data=subset(plot_df, marker == ''), aes(Z1, Z2), color='lightgray', size=0.25) +
  geom_point(color=alpha('black', 0), shape=21) +
  geom_point(inherit.aes=FALSE, data=subset(plot_df, anno != ''), aes(Z1,Z2, fill=rank), size=2, color='black', shape=21) +
  geom_text_repel(
    data = filter(plot_df, anno != ''),
    aes(label=anno),
    min.segment.length=0,
    fontface='italic', color='black', size=3,
    max.overlaps=1000
  ) +
  scale_fill_gradientn(
    colors=plasma(256),
    guide = guide_colorbar(barwidth=0.5, barheight=5, ticks=FALSE),
    breaks = range(plot_df$rank),
    labels = c('-', '+'),
  ) + labs(color='') + ggtitle('Gene Trajectory Rank') + theme(plot.title=element_text(hjust=0.5))


pdf(paste0(fig_dir,  cur_celltype, '_expression_Z_rank_markers.pdf'), width=5, height=5, useDingbats=FALSE)
p + umap_theme
dev.off()



################################################################################
# save useful variables
################################################################################

save(z_df, recon_df, ordered, file=paste0(data_dir, cur_celltype, '_expression_trajectory_data.rda'))


```



Plot reconstructed accessibility heatmap:

```{r eval=FALSE}

library(ComplexHeatmap)
library(future.apply)
library(viridis)

# RVAE directory
outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
cur_name <- paste0(cur_celltype, '_RVAE_tDARs')

# load reconstructed data and z embedding
dataset = cur_name
recon_df <- read.table(paste0(outdir, 'reconstructed/',cur_name,'_recon.csv'), sep=',', row.names=1, header=TRUE)
z_df <- recon_df[,(ncol(recon_df)-1):ncol(recon_df)]
recon_df <- recon_df[,1:(ncol(recon_df)-2)]
z_df$peak <- rownames(z_df)

# get promoter & gl-ccre info
promoter_info <- peakset %>% subset(peakType == 'Promoter') %>% dplyr::rename(gene = nearestGene)
glccre_info <- link_df %>% subset(Peak2 %in% z_df$peak) %>% select(c(Peak2, Peak2_type, Peak1_nearestGene)) %>% dplyr::rename(c(peak = Peak2, peakType = Peak2_type, gene = Peak1_nearestGene))


glccre_info <- data.frame(
  peak = unique(glccre_info$peak),
  gene = sapply(unique(glccre_info$peak), function(x){
    cur_genes <- subset(glccre_info, peak == x) %>% .$gene %>% unique
    paste(cur_genes, collapse=',')
  })
)
glccre_info$peakType <- link_df[match(glccre_info$peak, link_df$Peak2),'Peak2_type']
glccre_info <- glccre_info %>% select(c(peak, peakType, gene))

peak_info <- rbind(promoter_info, glccre_info)

# add info to z_df
z_df <- cbind(
  z_df,
  peak_info[match(z_df$peak, peak_info$peak),c('peakType', 'gene')]
)
z_df$peakType <- factor(as.character(z_df$peakType), levels=c('Promoter', 'Distal', 'Exonic', 'Intronic'))

# scale between 0 & 1
range01 <- function(x){
  cur <- recon_df[x,]
  (cur-min(cur))/(max(cur)-min(cur))
}
scaled <- lapply(rownames(recon_df), range01)
scaled <- do.call(rbind, scaled)
rownames(scaled) <- rownames(recon_df)

# order rows by the time they reach 0.75% min expression
ordering <- future_lapply(1:nrow(scaled), function(i){
  match(names(scaled[i,])[scaled[i,] >= 0.75][1], colnames(scaled))
})
ordered <- scaled[rownames(scaled)[order(unlist(ordering))],]

# reverse ordering
ordered <- ordered[rownames(ordered),]
z_df$rank <- match(rownames(z_df), rownames(ordered))

# order z_df by ranking
z_df <- z_df %>% arrange(rank)

# for oligos label important ODC lineage marker genes (from Marques et al 2016):
odc_lineage_genes <- c(
  'Pdgfra', 'Cspg4',
  'Vcan', 'Sox6', 'Nkx2-2',
  'Tcf7l2', 'Itpr2', 'Tmem2', 'Pdgfa',
  'Mal', 'Mog', 'Plp1', 'Opalin', 'Mbp',
  'Klk6', 'Apod'
)
cur_anno_genes <- odc_lineage_genes


n_degs <- 15
plot_genes <- markers %>%
  arrange(group) %>%
  subset(p_val_adj <= 0.05) %>%
  group_by(group) %>%
  top_n(n_degs, wt=avg_log2FC)  %>%
  .$gene %>% unique
odc_lineage_genes <- plot_genes
cur_anno_genes <- odc_lineage_genes

# using all target genes for the gl-cCRE as labels
# z_df$marker <- sapply(z_df$gene, function(x){
#   as.logical(length(intersect(odc_lineage_genes, unlist(strsplit(x, ',')))))
# })

# only labeling the marker gene
z_df$marker <- sapply(z_df$gene, function(x){
  cur_int = intersect(odc_lineage_genes, unlist(strsplit(x, ',')))
  if(length(cur_int) >= 1){
    out = cur_int
  } else{
    out = ''
  }
  out
})


# label diagnosis cluster DEGs

degs_indices <- (1:nrow(z_df))[z_df$marker != '']
degs_labels = z_df$marker[degs_indices]
degs_colors = rep('blue', length(degs_labels))
names(degs_colors) <- degs_labels
ha = rowAnnotation(
  foo = anno_mark(
    at = degs_indices, labels = degs_labels
), col=list(foo=degs_colors))


# split into promoter & gl-cCREs:
promoter_peaks <- proj@peakSet %>% subset(peakType == 'Promoter') %>% .$peak %>% unique
glccre_peaks <- link_df$Peak2 %>% unique %>% as.character

ordered_glccres <- ordered[rownames(ordered) %in% glccre_peaks,]
ordered_promoters <- ordered[rownames(ordered) %in% promoter_peaks,]

# plot heatmap
colfunc.atac <- colorRampPalette(c(brewer.pal(9, 'RdPu' )[2:9]))


pdf(paste0(fig_dir, cur_celltype, '_tDAR_recon_heatmap.pdf'), width=9, height=18)
ComplexHeatmap::Heatmap(
  as.matrix(ordered),
  show_column_names = FALSE, show_row_names=FALSE,
  col = colfunc.atac(256),
  cluster_rows=FALSE,
  cluster_columns=FALSE,
  right_annotation=ha,
  row_split = z_df$peakType,
  #left_annotation=color_ha,
  use_raster = TRUE
)
dev.off()


################################################################################
# plot latent space colored by pseudotime rank
################################################################################
library(ggrepel)

set.seed(42)

theme_set(theme_cowplot())

z_df$rank <- match(rownames(z_df), rownames(ordered))

# annotate marker genes
plot_df <- z_df
plot_df$anno <- ifelse(plot_df$marker %in% odc_lineage_genes, plot_df$marker, '')

# only allow the same marker gene to show up so many times:
max_markers <- 2
unique_markers <- unique(plot_df$anno)
unique_markers <- unique_markers[unique_markers != '']
cur <- unique_markers[1]

anno_df <- data.frame()
for(cur in unique_markers){
  cur_df <- plot_df %>% subset(anno == cur)
  if(nrow(cur_df) > max_markers){
    cur_df[sample(1:nrow(cur_df), nrow(cur_df) - max_markers), 'anno'] <- ''
  }
  anno_df <- rbind(anno_df, cur_df)
}

plot_df <- rbind(
  anno_df,
  subset(plot_df, anno == '')
)


# plot genes that are cell-type markers
p <- plot_df %>%
  ggplot(aes(Z1, Z2, color=rank)) +
#  geom_point(data=subset(plot_df, marker == ''), aes(Z1, Z2), color='lightgray', size=0.25) +
  geom_point() +
#  geom_text_repel(data = filter(plot_df, anno != ''), aes(label=anno), min.segment.length=0, fontface='italic', color='black') +
  scale_color_gradientn(
    colors=plasma(256),
    guide = guide_colorbar(barwidth=0.5, barheight=5, ticks=FALSE),
    breaks = range(plot_df$rank),
    labels = c('-', '+'),
  ) + labs(color='') + ggtitle('Accessibility Trajectory Rank') + theme(plot.title=element_text(hjust=0.5))

pdf(paste0(fig_dir,  cur_celltype, '_accessibility_Z_rank_unlabeled.pdf'), width=5, height=5, useDingbats=FALSE)
p + umap_theme
dev.off()


p <- plot_df  %>%
  ggplot(aes(Z1, Z2, fill=rank)) +
  geom_point(color=alpha('black', 0), shape=21) +
  geom_point(inherit.aes=FALSE, data=subset(plot_df, anno != ''), aes(Z1,Z2, fill=rank), size=2, color='black', shape=21) +
  geom_text_repel(
    data = filter(plot_df, anno != ''),
    aes(label=anno),
    min.segment.length=0,
    fontface='italic', color='black', size=3,
    max.overlaps=1000
  ) +
  scale_fill_gradientn(
    colors=plasma(256),
    guide = guide_colorbar(barwidth=0.5, barheight=5, ticks=FALSE),
    breaks = range(plot_df$rank),
    labels = c('-', '+'),
  ) + labs(color='') + ggtitle('Accessibility Trajectory Rank') + theme(plot.title=element_text(hjust=0.5))


pdf(paste0(fig_dir,  cur_celltype, '_accessibility_Z_rank_markers.pdf'), width=5, height=5, useDingbats=FALSE)
p + umap_theme
dev.off()

'Klk6' %in% z_df$marker

################################################################################
# save useful variables
################################################################################

save(z_df, recon_df, ordered, file=paste0(data_dir, cur_celltype, '_accessibility_trajectory_data.rda'))


```



Correlation of reconstructed expression with proportion timepoints

* Need to run proportion section before doing this one

```{r eval=FALSE}


################################################################################
# plot expression latent space colored by pseudotime rank
################################################################################

# re-load z_df
load(file=paste0(data_dir, cur_celltype, '_expression_trajectory_data.rda'))

theme_set(theme_cowplot())

dir.create(paste0(fig_dir, 'Diagnosis_corr/'))

plot_list <- list(); i = 1
for(gr in c('Young', 'Old')){
  for(timepoint in c("Naive","Day.5", "Day.14", "Day.30" )){

    # gr <- 'Old'
    # timepoint = 'Day.5'
    #
    print(paste(gr, timepoint))

    prop_cor <- sapply(1:nrow(recon_df), function(i){
      cor(as.numeric(recon_df[i,]), subset(prop_df, Diagnosis == timepoint & group == gr) %>% .$proportion)
    })


    plot_df <- z_df
    plot_df$value <- prop_cor
    plot_df$marker <- ifelse(plot_df$gene %in% oligo_markers, plot_df$gene, '')

    p <- plot_df %>% subset(marker != '') %>%
      ggplot(aes(Z1, Z2, color=value)) +
      geom_point(data=subset(plot_df, marker == ''), aes(Z1, Z2), color='lightgray', size=0.2) +
      geom_point(size=1) +
      scale_color_gradient2(
        low=scales::muted('blue'), mid='gray95', high=scales::muted('red'),
        guide = guide_colorbar(barwidth=0.5, barheight=10, ticks=FALSE)
      ) +
       umap_theme +
      theme(plot.title = element_text(hjust=0.5)) +
      # ggtitle('')
      ggtitle(paste0(gr, ', ', gsub('[.]', ' ', timepoint)))

      pdf(paste0(fig_dir, 'Diagnosis_corr/', gr, '_', timepoint, '_corr_diagnosis.pdf'), width=5, height=5, useDingbats=FALSE)
      print(p)
      dev.off()

    plot_list[[i]] <- p + NoLegend()
    i <- i+1
  }
}


pdf(paste0(fig_dir, cur_celltype, '_expression_Z_correlation_diagnosis.pdf'), width=8, height=5.3333, useDingbats=FALSE)
wrap_plots(plot_list, ncol=4)
#plot_list[[1]]
#p
dev.off()

png(paste0(fig_dir, cur_celltype, '_expression_Z_correlation_diagnosis.png'), width=8, height=5.3333, res=500, units='in')
wrap_plots(plot_list, ncol=4)
dev.off()

################################################################################
# plot accessibility latent space colored by pseudotime rank
################################################################################


# re-load z_df
load(file=paste0(data_dir, cur_celltype, '_accessibility_trajectory_data.rda'))
z_df <- z_df[rownames(recon_df),]

plot_list <- list(); i = 1
for(gr in c('Young', 'Old')){
  for(timepoint in c("Naive","Day.5", "Day.14", "Day.30" )){

    print(paste(gr, timepoint))
    # gr <- 'Old'
    # timepoint = 'Day.5'
    #
    prop_cor <- sapply(1:nrow(recon_df), function(i){
      cor(as.numeric(recon_df[i,]), subset(prop_df, Diagnosis == timepoint & group == gr) %>% .$proportion)
    })


    plot_df <- z_df
    plot_df$value <- prop_cor
    #plot_df$marker <- ifelse(plot_df$gene %in% oligo_markers, plot_df$gene, '')

    p <- plot_df %>%
      ggplot(aes(Z1, Z2, color=value)) +
      geom_point(data=subset(plot_df, marker == ''), aes(Z1, Z2), color='lightgray', size=0.2) +
      geom_point(size=1) +
      scale_color_gradient2(
        low='darkorchid4', mid='gray95', high='seagreen',
        guide = guide_colorbar(barwidth=0.5, barheight=10, ticks=FALSE)
      ) +
       umap_theme +
      theme(plot.title = element_text(hjust=0.5)) +
      ggtitle('')

      pdf(paste0(fig_dir, 'Diagnosis_corr/', gr, '_', timepoint, '_accessibility_corr_diagnosis.pdf'), width=5, height=5, useDingbats=FALSE)
      print(p)
      dev.off()

    plot_list[[i]] <- p + NoLegend()
    i <- i+1
  }
}


png(paste0(fig_dir, cur_celltype, '_accessibility_Z_correlation_diagnosis.png'), width=8, height=5.3333, res=500, units='in')
wrap_plots(plot_list, ncol=4)
dev.off()


```


Correlate expression and chromvar activity of each TF with the pseudotime trajectory
to identify which TFs might be regulating genes at different points along
pseudotime

```{r eval=FALSE}

library(future.apply)
library(ggrepel)
plan(multicore, workers=8)

################################################################################
# Data setup
################################################################################

# load expression data (z_df, recon_df, ordered)
load(file=paste0(data_dir, cur_celltype, '_expression_trajectory_data.rda'))

ensembl <- read.csv(
  '/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt',
  sep='\t'
)

# load mouse <-> human ortholog table (ensembl)
load('/dfs7/swaruplab/smorabit/resources/hg38_mm10_ortho.rda')

# get relevant motif info:
DefaultAssay(seurat_celltype) <- 'peaks'
motif_names <- GetMotifData(seurat_celltype, slot='motif.names')
motif_df <- data.frame(
  motif_id = names(motif_names),
  motif_name = as.character(motif_names)
)
motif_df$motif_name_human <- gsub("\\(.*)", "", as.character(motif_df$motif_name))
motif_df$motif_name_human <- gsub('::', ',', as.character(motif_df$motif_name_human))

# mouse <-> human motif names
motif_df$motif_name_mouse <- sapply(motif_df$motif_name_human, function(x){

  if(grepl(',', x)){
    x <- as.character(do.call(rbind, strsplit(x, ',')))
    x <- na.omit(as.character(ensembl[match(x, ensembl$Gene.name), 'Mouse.gene.name']))
    x <- paste(x, collapse=',')
  } else{
    x <- as.character(ensembl[match(x, ensembl$Gene.name), 'Mouse.gene.name'])
  }
  x
})

# fix one weird motif name
motif_df[motif_df$motif_name == 'EWSR1-FLI1', 'motif_name_mouse'] <- 'Ewsr1,Fli1'

################################################################################
# Chromvar enrichment correlation analysis
################################################################################

# get TF deviation matrix
DefaultAssay(seurat_celltype) <- 'chromvar'
Idents(seurat_celltype) <- seurat_celltype$pseudotime_bins_100
deviation_binned <- AverageExpression(seurat_celltype, assay='chromvar')$chromvar

# check that it matches the motif_df
all.equal(rownames(deviation_binned), motif_df$motif_id)

# compute correlation between TF enrichment and pseudotime
tf_cor <- future_sapply(1:nrow(deviation_binned), function(i){
  cor(as.numeric(deviation_binned[i,]), 1:ncol(deviation_binned))
})

motif_df$accessibility_cor <- tf_cor

################################################################################
# TF expression correlation analysis
################################################################################

# get the gene names for the mouse TFs:
mouse_motifs <- motif_df$motif_name_mouse[motif_df$motif_name_mouse != ""]
test <- lapply(mouse_motifs, function(x){
  if(grepl(',', x)){
    x <- as.character(unlist(do.call(rbind, strsplit(x, ','))))
  }
  x
})
mouse_motifs <- as.character(unique(unlist(test)))

# get TF expression matrix
Idents(seurat_subset) <- seurat_subset$pseudotime_bins_100
expression_binned <- AverageExpression(subset(seurat_subset, tech == 'rna'))$RNA

# subset by TFs
expression_binned <- expression_binned[rownames(expression_binned) %in% motif_df$motif_name_mouse, ]

# compute correlation between TF expression and pseudotime
tf_cor <- future_sapply(1:nrow(expression_binned), function(i){
  cor(as.numeric(expression_binned[i,]), 1:ncol(expression_binned))
})


# add to motif df
motif_df$expression_cor <- tf_cor[match(motif_df$motif_name_mouse, rownames(expression_binned))]

################################################################################
# Plot the TF enrichment correlations ranked as a barplot
################################################################################

# label the top & bottom 5
plot_df <- motif_df

labeled <- plot_df %>% top_n(7, accessibility_cor) %>% .$motif_name_mouse
labeled <- c(labeled, (plot_df %>% top_n(-5, accessibility_cor) %>% .$motif_name_mouse))

plot_df$anno <- ifelse(
  plot_df$motif_name_mouse %in% labeled,
  plot_df$motif_name_mouse,
  ''
)



p <- plot_df %>%
  ggplot(aes(x=reorder(motif_id, accessibility_cor), y=accessibility_cor, fill=accessibility_cor)) +
  geom_col(width=1) +
  geom_text_repel(data = dplyr::filter(plot_df, anno != ''), aes(label=anno), min.segment.length=0, fontface='italic') +
  scale_fill_gradient2(
    low='darkorchid4', mid='gray95', high='seagreen',
    guide = guide_colorbar(barwidth=0.5, barheight=10, ticks=FALSE)
  ) +
  theme(
    axis.text.x = element_blank(),
    axis.line.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank()
  ) + NoLegend() +
  ylab('Correlation of TF enrichment\nwith pseudotime')



pdf(paste0(fig_dir,  cur_celltype, '_TF_enrichment_trajectory_correlation.pdf'), width=7, height=3, useDingbats=FALSE)
p
dev.off()


################################################################################
# Plot the TF enrichment correlations ranked as a barplot
################################################################################

# label the top & bottom 5
plot_df <- motif_df %>% subset(!is.na(expression_cor))

labeled <- plot_df %>% top_n(5, expression_cor) %>% .$motif_name_mouse
labeled <- c(labeled, (plot_df %>% top_n(-5, expression_cor) %>% .$motif_name_mouse))

plot_df$anno <- ifelse(
  plot_df$motif_name_mouse %in% labeled,
  plot_df$motif_name_mouse,
  ''
)

p <- plot_df %>%
  ggplot(aes(x=reorder(motif_id, expression_cor), y=expression_cor, fill=expression_cor)) +
  # geom_bar(stat='identity') +
  geom_col(width=1) +
  geom_text_repel(data = dplyr::filter(plot_df, anno != ''), aes(label=anno), min.segment.length=0, fontface='italic') +
  scale_fill_gradient2(
    low='blue', mid='gray95', high='red',
    guide = guide_colorbar(barwidth=0.5, barheight=10, ticks=FALSE)
  ) +
  theme(
    axis.text.x = element_blank(),
    axis.line.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank()
  ) + NoLegend() +
  ylab('Correlation of TF expression\nwith pseudotime')



pdf(paste0(fig_dir,  cur_celltype, '_TF_expression_trajectory_correlation.pdf'), width=7, height=3, useDingbats=FALSE)
p
dev.off()



```


Plot trajectory of individual features along pseudotime in different
groups (young/old, timepoints)

* genes
* promoters / enhancers
* TF enrichments


Am going to need to modify code for when there are enhancers, and when
there is only one peak!!



```{r eval=FALSE}

# load expression data:
load(file=paste0(data_dir, cur_celltype, '_expression_trajectory_data.rda'))
z_df_exp <- z_df; recon_df_exp <- recon_df; ordered_exp <- ordered;

# load accessibility data
load(file=paste0(data_dir, cur_celltype, '_accessibility_trajectory_data.rda'))
z_df_acc <- z_df; recon_df_acc <- recon_df; ordered_acc <- ordered;


################################################################################
# extract averaged expression, accessibility, and TF enrichment matrices from seurat obj
################################################################################

# get accessibility matrix
DefaultAssay(seurat_subset) <- 'RNA'
Idents(seurat_subset) <- seurat_subset$pseudotime_bins_100
exp_mat <- AverageExpression(subset(seurat_subset, tech=='rna'), assay='RNA')$RNA

# get accessibility matrix
DefaultAssay(seurat_celltype) <- 'peaks'
Idents(seurat_celltype) <- seurat_celltype$pseudotime_bins_100
acc_mat <- AverageExpression(seurat_celltype, assay='peaks')$peaks

# get TF deviation matrix
DefaultAssay(seurat_celltype) <- 'chromvar'
Idents(seurat_celltype) <- seurat_celltype$pseudotime_bins_100
dev_mat <- AverageExpression(seurat_celltype, assay='chromvar')$chromvar

################################################################################
# setup for plotting
################################################################################

# set the gene of interest
cur_gene <- 'Olig1'


# pick a gene that is high in acc but low in exp

motif_df %>% top_n(-50, expression_cor) %>% arrange(-accessibility_cor)

cur_gene <- 'Sox10'


cur_gene <- 'Nkx2-2'
link_df[grepl('Nkx2-2', link_df$Peak1_nearestGene),]




# get promoter region(s) for this gene
cur_promoters <- peakset %>% subset(nearestGene == cur_gene & peakType == 'Promoter') %>% .$peak

# get enhancers for this gene (if there are any)
cur_enhancers <- link_df %>% subset(Peak1_nearestGene == cur_gene) %>% .$Peak2

# combine promoters & enhancers
cur_peaks <- c(cur_promoters, cur_enhancers)

# get the motif based on this gene:
cur_motif_id <- subset(motif_df, motif_name_mouse == cur_gene) %>% .$motif_id

print(cur_gene)
print(cur_peaks)
print(cur_motif_id)

# extract features from matrices:
cur_exp <- exp_mat[cur_gene,]
cur_acc <- acc_mat[cur_peaks,]
cur_dev <- dev_mat[cur_motif_id,]


# scale between 0 and 1
range01 <- function(x){
  (x-min(x))/(max(x)-min(x))
}

cur_exp <- range01(cur_exp)
cur_dev <- range01(cur_dev)
cur_acc <- apply(cur_acc, 1, range01) %>% t

################################################################################
# test plotting with geom_smooth
################################################################################

# initialize with gene info
plot_df <- data.frame(
  value = cur_exp,
  pseudotime = 1:length(cur_exp),
  FeatureType = 'Expression',
  feature = cur_gene
)

# add accessibility info:
for(i in 1:nrow(cur_acc)){
  feat_type <- peakset[peakset$peak == rownames(cur_acc)[i], 'peakType']
  cur_df <- data.frame(
    value = cur_acc[i,],
    pseudotime = 1:length(cur_exp),
    FeatureType = ifelse(feat_type == 'Promoter', 'Promoter', 'gl-cCRE'),
    feature = rownames(cur_acc)[i]
  )
  plot_df <- rbind(plot_df, cur_df)
}

# add motif info:
plot_df <- rbind(plot_df, data.frame(
  value = cur_dev,
  pseudotime = 1:length(cur_exp),
  FeatureType = 'Motif',
  feature = cur_motif_id
))


# re-order
plot_df$feature <- factor(
  plot_df$feature,
  levels = c(cur_gene, cur_motif_id, cur_promoters, cur_enhancers)
)


# make a plot!!!

n_feats <- length(unique(plot_df$feature))
h <- ceiling(n_feats / 5)

p <- plot_df %>%
  ggplot(aes(x=pseudotime, y=value, fill=FeatureType, color=FeatureType)) +
  geom_point(size=0.25, color='lightgray') +
  geom_smooth() +
  ylab('scaled value') +
  xlab('pseudotime') +
  theme(
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
  )


pdf(paste0(fig_dir, '/feature_trajectories/', cur_gene, '_test_feature_trajectory.pdf'), width=10, height=2*h, useDingbats=FALSE)
p + facet_wrap(~feature, ncol=5)
dev.off()


```


CoveragePlot and co-accessibility plots for gene of interest:

```{r eval=FALSE}



# get the coordinates
cur_region <- Annotation(seurat_celltype) %>% subset(gene_name == cur_gene)
cur_chr <- unique(as.character(seqnames(cur_region)))

################################################################################
# plot ideogram
################################################################################

cur_pos <- subset(genebodyandpromoter.coords, symbol==cur_gene)
chr <- as.character(seqnames(cur_pos))
gen <- 'mm10'
itrack <- IdeogramTrack(genome = gen, chromosome = chr)
print(paste(cur_gene, 'chr:', as.character(seqnames(cur_pos)), ',', abs(start(cur_pos) - end(cur_pos)) + 2000))

pdf(paste0(fig_dir, '/coverage_plots/', cur_celltype, '_', cur_gene, '_ideogram.pdf'), width=8, height=2, useDingbats=FALSE)
plotTracks(list(itrack), from=start(cur_pos), to=end(cur_pos), showId=FALSE)
dev.off()

################################################################################
# cicero connections
################################################################################

cur_pos <- subset(genebodyandpromoter.coords, symbol==cur_gene)

cur_conns <- link_df %>% subset(Peak1_nearestGene == cur_gene) %>%
  dplyr::select(c(Peak1, Peak2, coaccess))
rownames(cur_conns) <- 1:nrow(cur_conns)

cur_peak1_ranges <- peak1_ranges %>% subset(peak %in% cur_conns$Peak1)
cur_peak2_ranges <- peak2_ranges %>% subset(peak %in% cur_conns$Peak2)

cur_start <- min(start(cur_peak1_ranges), start(cur_peak2_ranges)) - 10000
cur_end <- max(end(cur_peak1_ranges), end(cur_peak2_ranges)) + 10000


view=paste0('chr',as.character(seqnames(cur_pos)), '_', start(cur_pos), '_', end(cur_pos))

pdf(paste0(fig_dir, '/coverage_plots/', cur_celltype, '_', cur_gene, '_coaccessibility.pdf'), width=5, height=3, useDingbats=FALSE)

cicero::plot_connections(
  cur_conns, cur_chr, cur_start, cur_end,
  gene_model = gene_anno,
  alpha_by_coaccess=TRUE,
  connection_color="black",
  coaccess_cutoff = 0,
  connection_width = 1,
  collapseTranscripts = "longest",
  #connection_ymax=coaccess_ymax,
  viewpoint=view,
#  return_as_list=TRUE
)

dev.off()


################################################################################
# Signac coverage plot
################################################################################

p <- CoveragePlot(
  object = seurat_celltype,
  group.by='pseudotime_bins_10',
  region = paste0(cur_chr, '-', cur_start, '-', cur_end),
  annotation = TRUE, peaks = FALSE, tile = FALSE, links = FALSE
)
pdf(paste0(fig_dir, '/coverage_plots/', cur_celltype, '_', cur_gene, '_pseudotime_coverage.pdf'), width=6, height=12, useDingbats=FALSE)
print(p)
dev.off()



#
# # plot cicero connections only, no peaks!
# pdf(paste0(fig_dir,cur_celltype,"_",cur_gene,"_diagnosis.pdf"), width=10, height=10)
# print(Gviz::plotTracks(
#    #trackList=list(p_AD@trackList[[1]], gwas_track, p_control@trackList[[1]], grtrack, p_AD@trackList[[3]]),
#    trackList=list(p_AD@trackList[[1]], cur_gwas_track, p_control@trackList[[1]], grtrack, p_AD@trackList[[3]]),
#    sizes = c(4,1,4,2,1),
#    from = cur_start, to = cur_end, chromosome = cur_chr,
#    transcriptAnnotation = "symbol",
#    col.axis = "black",
#    fontsize.group = 6,
#    fontcolor.legend = "black",
#    lwd=.3,
#    title.width = .5,
#    background.title = "transparent",
#    col.border.title = "transparent"
# ))
# dev.off()

```





Proportion of cell-types across pseudotime in Young & Old

```{r eval=FALSE}

library(ggstream)


theme_set(theme_pubr())

celltype_prop_df <- data.frame()
valid_pseudotime <- seurat_subset@meta.data %>% subset(pseudotime != Inf)

# Young
valid_pseudotime_age <- seurat_subset@meta.data %>% subset(pseudotime != Inf & Group == "Young")

# get proportion of cells in each bin for each time point
bins <- levels(valid_pseudotime_age$pseudotime_bins_100)
proportion_df <- data.frame()
for(i in 1:length(bins)){
  bin = bins[i]

  cur_meta <- subset(valid_pseudotime_age, pseudotime_bins_100 == bin)

  cur_df <- data.frame(table(cur_meta$subcluster_name) / nrow(cur_meta))
  cur_df <- cur_df %>% dplyr::rename(c(celltype=Var1, proportion=Freq))
  cur_df$bin_num <- i
  cur_df$bin <- bin
  proportion_df <- rbind(proportion_df, cur_df)

}

# set factor levels:
proportion_df$celltype <- factor(
  as.character(proportion_df$celltype),
  levels = c('OPC', 'NFOL', 'MF-ODC', 'Int-ODC', 'Mat-ODC')
)

proportion_df$celltype <- fct_rev(proportion_df$celltype)
proportion_df$group <- 'Young'
celltype_prop_df <- rbind(celltype_prop_df, proportion_df)

p1 <- ggplot(proportion_df, aes(x=bin_num, y=proportion, fill=celltype)) +
  xlab('Trajectory') + ylab('') +
  geom_stream(type='proportional') +
  scale_fill_manual(values=cluster_colors_rna[unique(seurat_subset$subcluster_name)]) +
  theme(
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) + umap_theme


pdf(paste0(fig_dir, cur_celltype, '_celltype_proportions_young_100_stream.pdf'), width=5, height=2, useDingbats=FALSE)
p1
dev.off()

# Old
valid_pseudotime_age <- seurat_subset@meta.data %>% subset(pseudotime != Inf & Group == "Old")

# get proportion of cells in each bin for each time point
bins <- levels(valid_pseudotime_age$pseudotime_bins_100)
proportion_df <- data.frame()
for(i in 1:length(bins)){
  bin = bins[i]

  cur_meta <- subset(valid_pseudotime_age, pseudotime_bins_100 == bin)

  cur_df <- data.frame(table(cur_meta$subcluster_name) / nrow(cur_meta))
  cur_df <- cur_df %>% dplyr::rename(c(celltype=Var1, proportion=Freq))
  cur_df$bin_num <- i
  cur_df$bin <- bin
  proportion_df <- rbind(proportion_df, cur_df)

}

# set factor levels:
proportion_df$celltype <- factor(
  as.character(proportion_df$celltype),
  levels = c('OPC', 'NFOL', 'MF-ODC', 'Int-ODC', 'Mat-ODC')
)

proportion_df$celltype <- fct_rev(proportion_df$celltype)
proportion_df$group <- 'Old'
celltype_prop_df <- rbind(celltype_prop_df, proportion_df)

p1 <- ggplot(proportion_df, aes(x=bin_num, y=proportion, fill=celltype)) +
  xlab('Trajectory') + ylab('') +
  geom_stream(type='proportional') +
  scale_fill_manual(values=cluster_colors_rna[unique(seurat_subset$subcluster_name)]) +
  theme(
    axis.text.x=element_blank(),
    axis.ticks.x=element_blank(),
    plot.title = element_text(hjust = 0.5)
  ) + umap_theme


pdf(paste0(fig_dir, cur_celltype, '_celltype_proportions_old_100_stream.pdf'), width=5, height=2, useDingbats=FALSE)
p1
dev.off()

```









Correlate TFs with expression  recon df (might move this section to a different notebook cause this one is getting so long ):
```{r eval=FALSE}

load(file=paste0(data_dir, cur_celltype, '_expression_trajectory_data.rda'))

# re-load expression recon_df



plan(multicore, workers=8)

Idents(seurat_celltype) <- seurat_celltype$pseudotime_bins_100
deviation_binned <- AverageExpression(seurat_celltype, assay='chromvar')$chromvar


# identify the top motifs for this celltype: ###################################
seurat_celltype <- FindTopFeatures(
  seurat_celltype,
  min.cutoff=ncol(seurat_celltype)*0.05 # present in 10% of cells
)
length(VariableFeatures(seurat_celltype))


motif_names <- GetMotifData(seurat_celltype, slot='motif.names')
motifs.use <- names(motif_names)
cur_motif_accessible <- Motifs(seurat_celltype)@data[proj@peakSet$peakType == 'Promoter' & rownames(seurat_celltype) %in% VariableFeatures(seurat_celltype),motifs.use]

# number of total binding sites for each TF:
bs_accessible <- Motifs(seurat_celltype)@data[proj@peakSet$peakType == 'Promoter',]
n_bs <- colSums(bs_accessible)
n_accessible_bs <- colSums(cur_motif_accessible)
df <- data.frame(value = n_accessible_bs / n_bs)
df$percent <- df$value * 100
df$motif_name <- as.character(motif_names)

# only take top motifs:
cur_motif_accessible <- cur_motif_accessible[,rownames(df)[df$value >= quantile(df$value, 0.65)]]

#subset deviation binned to keep these motifs
motifs.keep <- colnames(cur_motif_accessible)
deviation_binned_full <- deviation_binned
deviation_binned <- deviation_binned[motifs.keep,]
cur_motif_names <- as.character(motif_names[motifs.keep])
cur_motif_names <- as.character(motif_names)
deviation_binned <- deviation_binned_full

# compute correlation between gene exp and TF enrichment
cor_mat <- future_sapply(1:nrow(recon_df), function(i){
  print(i)
  sapply(1:nrow(deviation_binned), function(j){
  cor(as.numeric(recon_df[i,]), as.numeric(deviation_binned[j,]))
})})


colnames(cor_mat) <- rownames(recon_df)
rownames(cor_mat) <- as.character(motif_names)
cor_mat <- t(cor_mat)

# order genes
cor_mat <- cor_mat[rownames(ordered),]

# save correlation matrix:
write.csv(cor_mat, paste0('data/', cur_celltype, '_TF_VAE_correlation.csv'), quote=FALSE )

# annotate motif names (celltype marker TFs & up/down in AD):
pdf(paste0(fig_dir, cur_celltype, '_TF_corr_heatmap.pdf'), width=4, height=8)
ComplexHeatmap::Heatmap(
  cor_mat,
  show_column_names = FALSE, show_row_names=FALSE,
  col = rev(brewer.pal(11, 'Spectral')),
  cluster_rows=FALSE,
  cluster_columns=TRUE,
  use_raster=TRUE
)
dev.off()


# plot TF correlations on the Z-embedding ######################################

# re-load corr mat:
cor_mat <- read.csv(paste0('data/', cur_celltype, '_TF_VAE_correlation.csv'), row.names=1)

dir.create(paste0(fig_dir, '/TF_correlations'))

library(ggrepel)
library(tidyverse)

# load DA motifs
# motif_names <- GetMotifData(seurat_celltype, slot='motif.names')
# motifs.use <- names(motif_names)
# load('/dfs7/swaruplab/smorabit/analysis/AD_NucSeq_2019/atac_analysis/all_data/chromVar/data/diagnosis_da_motifs.rda')
# load('/dfs7/swaruplab/smorabit/analysis/AD_NucSeq_2019/atac_analysis/all_data/chromVar/data/da_motifs.rda')
# da_motifs_celltypes$motif_name <- motif_names[da_motifs_celltypes$gene]
# diagnosis_da_motifs$motif_name <- motif_names[diagnosis_da_motifs$gene]

# select motifs:
#tf_list <- c('SPI1', 'ETS1', 'FOXO3')
# other_tfs <- c('STAT3') # ASC
# other_tfs <- c()
# other_tfs <- c('SREBF1')
#
# tf_list_celltype <- da_motifs_celltypes %>% subset(cluster==cur_celltype & avg_logFC >= 0) %>%
#   top_n(20, wt=avg_logFC) %>% .$motif_name %>% as.character
#
# tf_list_up_AD <- diagnosis_da_motifs %>%
#   subset(cluster=='AD' & celltype==cur_celltype & avg_logFC >=0) %>%
#   top_n(10, wt=avg_logFC) %>% .$motif_name %>% as.character
#
# tf_list_down_AD <- diagnosis_da_motifs %>%
#   subset(cluster=='AD' & celltype==cur_celltype & avg_logFC <0) %>%
#   top_n(10, wt=-avg_logFC) %>% .$motif_name %>% as.character
#
# tf_list <- c(tf_list_celltype, tf_list_up_AD, tf_list_down_AD, other_tfs) %>% unique

tf_list <- c("SREBF1", 'SREBF2', 'OLIG1', 'OLIG2', 'SOX9', 'SOX10', 'NKX2-2', 'NRF1')
unlist(motif_names)[grep('NKX', unlist(motif_names))]

# get matrix of motif binding sites
cur_promoter_motif_accessible <- Motifs(seurat_celltype)@data[proj@peakSet$peakType == 'Promoter' ,]
colnames(cur_promoter_motif_accessible) <- as.character(motif_names)
# cur_linked_motif_accessible <- Motifs(seurat_celltype)@data[proj@peakSet$site_name2 %in% cur_links$Peak2,]
# colnames(cur_linked_motif_accessible) <- as.character(motif_names)

# what genes have a binding site
plot_list <- list(); i <- 1;
plot_df <- data.frame()
for(cur_tf in tf_list){
  print(cur_tf)
  cur_df <- z_df

  # which genes have a binding site for this TF?
  cur_promoter_bs <- rownames(cur_promoter_motif_accessible)[cur_promoter_motif_accessible[,cur_tf] > 0]

  # which genes have a link to an enhancer with a binding site to this TF?
  # cur_enhancer_bs <- rownames(cur_linked_motif_accessible)[cur_linked_motif_accessible[,cur_tf] >0]

  # what genes do these correpsond to?
  cur_promoter_bs_genes <- subset(proj@peakSet, site_name %in% cur_promoter_bs) %>% .$nearestGene %>% as.character
  # cur_enhancer_bs_genes <- subset(proj@peakSet, site_name %in% cur_enhancer_bs) %>% .$nearestGene %>% as.character

  # cur_bs_genes <- c(cur_promoter_bs_genes, cur_enhancer_bs_genes) %>% unique
  cur_bs_genes <- c(cur_promoter_bs_genes) %>% unique
  cur_enhancer_bs_genes <- c()

  cur_df$bs <- cur_df$gene %in% cur_bs_genes

  cur_df$bs_group <- ifelse(!cur_df$bs, NA, ifelse(
    cur_df$gene %in% cur_promoter_bs_genes & cur_df$gene %in% cur_enhancer_bs_genes, 'both', ifelse(
    cur_df$gene %in% cur_promoter_bs_genes, 'promoter', ifelse(
      cur_df$gene %in% cur_enhancer_bs_genes, 'enhancer', NA
  ))))


  if(sum(na.omit(cur_df$bs_group) == 'both') == 0){
    shapes = c(22,21)
  } else{
    shapes = c(24,22,21)
  }

  # make all shapes 21 (remove later)
  shapes <- 21

  # cur_df$bs_group <- factor(cur_df$bs_group, levels=c('both', 'enhancer', 'promoter'))

  # label certain genes of interest?
  cur_df$anno <- ifelse(cur_df$gene %in% odc_lineage_genes & cur_df$bs, cur_df$gene, '')


  cur_df$variable <- cur_tf
  cur_df$value <- as.numeric(cor_mat[z_df$rank, cur_tf])
  plot_df <- rbind(plot_df, cur_df)

  # limit of color scale:
  limits <- max(abs(min(cur_df$value)), max(cur_df$value))
  limits <- c(-1*limits, limits)

#  p <- ggplot(subset(cur_df, bs), aes(Z1, Z2, fill=value, shape=bs_group)) +
  p <- ggplot(subset(cur_df, bs), aes(Z1, Z2, fill=value, shape=bs_group)) +
    geom_point(inherit.aes=FALSE, data=subset(cur_df, !bs), aes(Z1,Z2), color='gray', fill='gray', alpha=0.25, size=0.5) +
    geom_point(size=2, color=alpha('black', 0)) +
  #  geom_point(inherit.aes=FALSE, data=subset(cur_df, bs & !is.na(anno)), aes(Z1,Z2, fill=value, shape=bs_group), size=2, color='black') +
    scale_shape_manual(values=shapes, guide=FALSE) +
    geom_text_repel(data = filter(cur_df, anno != ''), aes(label=anno), min.segment.length=0, fontface='italic') +
    scale_fill_gradientn(
      colors=rev(brewer.pal(11, 'Spectral')),
      limit = limits,
      breaks = c(limits[1], 0, limits[2]),
      labels = c('-', 0, '+'),
      guide = guide_colorbar(barwidth=0.5, barheight=5, ticks=FALSE)
    ) +
  #  scale_color_manual(values=subset(cur_df, bs) %>% .$anno_outline) +
    labs(fill='',shape='') +
    ggtitle(paste0(cur_tf)) + theme(plot.title = element_text(hjust=0.5))


  plot_list[[i]] <- p + umap_theme + NoLegend()
  i <- i + 1

  pdf(paste0(fig_dir, '/TF_correlations/', cur_celltype, '_Z_', cur_tf,'_correlation.pdf'), width=5, height=5, useDingbats=FALSE)
  print(p  + umap_theme)
  dev.off()

}

# combined plot
pdf(paste0(fig_dir, '/TF_correlations/', cur_celltype, '_Z_combined.pdf'), width=10, height=6, useDingbats=FALSE)
wrap_plots(plot_list, ncol=4)
dev.off()


sum(plot_df$anno_group !='white')
sum(!is.na(plot_df$anno))


```
