
```{r eval=FALSE}


# conda activate cicero
library(Seurat)
library(Signac)
library(tidyverse)
library(RColorBrewer)
library(viridis)
library(cowplot)
library(ggrepel)
library(ggpubr)
library(monocle3)
theme_set(theme_cowplot())

# scp R/* hpc3:/dfs7/swaruplab/smorabit/analysis/galilean/bin/

scripts <- dir("/dfs7/swaruplab/smorabit/analysis/galilean/bin/")
for(script in scripts){
  source(paste0("/dfs7/swaruplab/smorabit/analysis/galilean/bin/", script))
}

set.seed(12345)
setwd("/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC")

data_dir <- 'data/'
fig_dir <- 'figures/'

umap_theme <- theme(
  axis.line=element_blank(),
  axis.text.x=element_blank(),
  axis.text.y=element_blank(),
  axis.ticks=element_blank(),
  axis.title.x=element_blank(),
  axis.title.y=element_blank(),
  panel.background=element_blank(),
  panel.border=element_blank(),
  panel.grid.major=element_blank(),
  panel.grid.minor=element_blank()
)

# set the current cell-type (oligos)
cur_celltype <- 'ODC'

fig_dir <- paste0('figures/', cur_celltype, '/'); dir.create(fig_dir)


################################################################################
# load Seurat objects
################################################################################

# load the color scheme:
load(file='/dfs7/swaruplab/smorabit/collab/AMRF/data/color_scheme.rda')

# load the co-embedding object for the oligo lineage:
seurat_obj <- readRDS('/dfs7/swaruplab/smorabit/collab/AMRF/analysis//ATAC/data/ODC_coembed_seurat.rds')

# load the ATAC & RNA objects separately, and subset just by the oligo lineage cells:
seurat_atac <- readRDS('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/signac_final_07-02-21.rds')
DefaultAssay(seurat_atac) <- 'peaks'
seurat_atac$Time.point <- seurat_atac$Timepoint

seurat_rna <- readRDS("/dfs7/swaruplab/smorabit/collab/AMRF/analysis/snRNAseq/NucSeq_final_doublets_removed_online_inmf_seurat.rds")

atac_ix <- na.omit(match(colnames(seurat_obj), colnames(seurat_atac)))
rna_ix <- na.omit(match(colnames(seurat_obj), colnames(seurat_rna)))

# subset by atac/rna cells that are in the oligo pseudotime object:
seurat_atac <- seurat_atac[,atac_ix]
seurat_rna <- seurat_rna[,rna_ix]

# add the co-embed umap to atac/rna objects:
atac_ix <- na.omit(match(colnames(seurat_atac), colnames(seurat_obj)))
rna_ix <- na.omit(match(colnames(seurat_rna), colnames(seurat_obj)))
atac_umap <- seurat_obj@reductions$umap@cell.embeddings[atac_ix,]
rna_umap <- seurat_obj@reductions$umap@cell.embeddings[rna_ix,]
seurat_rna@meta.data <- cbind(seurat_rna@meta.data, seurat_obj@meta.data[rna_ix,c('pseudotime', 'monocle_clusters', 'monocle_partitions', 'pseudotime_bins_10', 'pseudotime_bins_50', 'pseudotime_bins_100')])
seurat_atac@meta.data <- cbind(seurat_atac@meta.data, seurat_obj@meta.data[atac_ix,c('pseudotime', 'monocle_clusters', 'monocle_partitions', 'pseudotime_bins_10', 'pseudotime_bins_50', 'pseudotime_bins_100')])

seurat_atac@reductions$seurat_obj_umap <- CreateDimReducObject(
  embeddings = atac_umap,
  key='UMAP_',
  assay='RNA'
)

seurat_rna@reductions$seurat_obj_umap <- CreateDimReducObject(
  embeddings = rna_umap,
  key='UMAP_',
  assay='RNA'
)

seurat_obj$subcluster_name <- factor(
  as.character(seurat_obj$subcluster_name),
  levels = c('OPC', 'NFOL', 'MF-ODC', 'Int-ODC', 'Mat-ODC')
)

```

subsetting:

```{r eval=FALSE}

# cut pseudotime into bins:
#seurat_obj <- BinPseudotime(seurat_obj, n_bins=c(10,50,100))

#table(seurat_obj$Time.point, seurat_obj$pseudotime_bins_50)

# subset old/young:
cur_group <- 'Young'
cur_group <- 'Old'


subset.cells <- ((seurat_obj$Time.point == 'Naive' & seurat_obj$subcluster_name == 'OPC') |
(seurat_obj$Time.point == 'Day.5' & seurat_obj$subcluster_name %in% c('OPC', 'NFOL')) |
(seurat_obj$Time.point == 'Day.14' & seurat_obj$subcluster_name %in% c('MF-ODC', 'Int-ODC')) |
(seurat_obj$Time.point == 'Day.30' & seurat_obj$subcluster_name == 'Mat-ODC')) &
(seurat_obj$Group == cur_group)
table(subset.cells)

# subset the seurat obj
seurat_subset <- seurat_obj[,subset.cells]

# cut pseudotime into bins:
seurat_subset <- BinPseudotime(seurat_subset, n_bins=c(10,50,100))

atac_ix <- na.omit(match(colnames(seurat_subset), colnames(seurat_atac)))
rna_ix <- na.omit(match(colnames(seurat_subset), colnames(seurat_rna)))

seurat_atac_subset <- seurat_atac[,atac_ix]
seurat_rna_subset <- seurat_rna[,rna_ix]




```

stream plot:

```{r eval=FALSE}

library(ggstream)
library(patchwork)

# compute the proportions
seurat_subset  <- BinProportions(
  seurat_subset, 'pseudotime_bins_50', c('Time.point', 'subcluster_name')
)

# make proportion plots
pl <- PlotBinProportions(
  seurat_subset,
  meta = c('Time.point', 'subcluster_name'),
  pseudotime_bins = 'pseudotime_bins_50',
  arrow_y = -0.2,
  clean=TRUE,
  combined=FALSE
)


pl[[1]] <- pl[[1]] + scale_fill_brewer(palette='PuBu', direction=-1)
pl[[2]] <- pl[[2]] + scale_fill_manual(values=cluster_colors_rna[as.character(unique(seurat_subset$subcluster_name))])

pdf(paste0(fig_dir, cur_celltype, '_', cur_group, '_proportion.pdf'), width=6, height=3)
wrap_plots(pl, ncol=1) + plot_layout(guides='collect')
dev.off()

```


Setup motif data (important for later steps)

```{r eval=FALSE}

# get list of TFs from atac data
DefaultAssay(seurat_atac) <- 'peaks'
chromvar_motifs <- Motifs(seurat_atac)@motif.names %>%
  unlist %>% as.character
print(length(chromvar_motifs))

# set up motif df
motif_df <- data.frame(
  motif_id = names(Motifs(seurat_atac)@motif.names),
  motif_name_orig = as.character(Motifs(seurat_atac)@motif.names)
)

# 1. split apart cases where two genes are one motif delimited by ::
# 2. remove "()" characters and anything within them

tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
tmp_ids <- names(Motifs(seurat_atac)@motif.names[chromvar_motifs %in% tmp])

motif_df$motif_name <- gsub("\\s*\\([^\\)]+\\)","", motif_df$motif_name_orig)
tmp <- motif_df$motif_name[grepl("::", motif_df$motif_name)]
motif_df <- subset(motif_df, !(motif_name %in% tmp))

tmp2 <- strsplit(tmp, '::')
names(tmp2) <- tmp
tmp <- do.call(rbind, lapply(1:length(tmp2), function(i){
  data.frame(motif_id = tmp_ids[i], motif_name_orig = names(tmp2)[i], motif_name = unlist(as.character(tmp2[[i]])))
}))
motif_df <- rbind(motif_df, tmp)

# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))

ix <- match(motif_df$motif_name, hg38_mm10_genes$hg38_name)
motif_df$motif_name_mouse <- hg38_mm10_genes$mm10_name[ix]


chromvar_motifs <- unique(motif_df$motif_name_mouse)

# which do we also have RNA data for?
chromvar_motifs <- chromvar_motifs[chromvar_motifs %in% rownames(seurat_rna)]

```

Calculate t-DEGs:

```{r eval=FALSE}

# make a new CDS with just the RNA-seq data in order to do t-DEGs:
expr_matrix  <- GetAssayData(subset(seurat_subset, tech == 'rna'), assay='RNA', slot='data')
pseudotime_genes <- rownames(expr_matrix )[rowSums(expr_matrix )!=0]
expr_matrix <- expr_matrix[pseudotime_genes,]

cds_subset <- new_cell_data_set(
  expr_matrix,
  cell_metadata=seurat_subset@meta.data %>% subset(tech == 'rna')
)

# identify t-DEGs
gene_fits <- monocle3::fit_models(
  cds_subset,
  model_formula_str = "~pseudotime + batch_integrated + nCount_RNA + percent.mt"
)

fit_coefs <- coefficient_table(gene_fits)

# save the full coefficient table:
tdegs <- fit_coefs
save(fit_coefs, file=paste0(data_dir, cur_celltype, '_', cur_group, '_tDEGs.rda'))


df <- seurat_subset@meta.data %>%
  subset(tech == 'rna') %>%
  select(c(pseudotime, batch_integrated, nCount_RNA, percent.mt))

df <- cbind(df, expr_matrix['Xkr4',])
colnames(df)[ncol(df)] <- 'Xkr4'

  as.data.frame(expr_matrix['Xkr4',])

#
predict(tdegs$model[1]$Xkr4, newdata=df)

tdegs <- tdegs %>%select(gene_id, term, q_value, estimate)

# write t-degs table as a csv:
write.csv(tdegs, file=paste0(data_dir, cur_celltype, '_', cur_group, '_tDEGs.csv'), quote=FALSE, row.names=FALSE)

# re-load t-degs:
tdegs <- read.csv(paste0(data_dir, cur_celltype, '_', cur_group, '_tDEGs.csv'))

# add t-degs to Seurat object:
seurat_subset <- SetDynamicFeatures(
  seurat_subset,
  feature_name = 'tDEGs',
  dynamic_features = tdegs,
  overwrite=TRUE
)

```


Which genes overlap between Young t-DEGs & old t-DEGs?

```{r eval=FALSE}

o_tdegs <- read.csv('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC_Old_tDEGs.csv')
y_tdegs <- read.csv('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC_Young_tDEGs.csv')

tdegs <- o_tdegs
tdegs <- y_tdegs

length(intersect(o_tdegs$gene_id, y_tdegs$gene_id))


library(eulerr)

overlap_list <- c(
  'Old' = length(setdiff(o_tdegs$gene_id, y_tdegs$gene_id)),
  'Young' =  length(setdiff(y_tdegs$gene_id, o_tdegs$gene_id)),
  'Old&Young' =  length(unique(intersect(o_tdegs$gene_id, y_tdegs$gene_id)))
)


pdf(paste0(fig_dir, 't-DEG_age_overlaps.pdf'), width=4, height=4)
plot(euler(overlap_list), quantities = TRUE)
dev.off()

###################################################################
# run enrichR on the genes that are unique to old & unique to young:
###################################################################



library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021','WikiPathway_2021_Human', 'KEGG_2021_Human')


old_only <- setdiff(o_tdegs$gene_id, y_tdegs$gene_id)
young_only <- setdiff(y_tdegs$gene_id, o_tdegs$gene_id)


enriched_old <- enrichR::enrichr(old_only, dbs)
enriched_young <- enrichR::enrichr(young_only, dbs)
enriched_list <- list('Old' = enriched_old, 'Young' = enriched_young)

# collapse into one db
combined_output <- data.frame()
for(group in names(enriched_list)){
  enriched <- enriched_list[[group]]
  for(db in names(enriched)){
    cur_df <- enriched[[db]]
    if (nrow(cur_df) > 1){
      cur_df$db <- db
      cur_df$group <- group
      combined_output <- rbind(combined_output, cur_df)
    }
  }
}

combined_output$ngenes <- unlist(lapply(strsplit(combined_output$Genes, ';'), function(x){length(x)}))
combined_output <- combined_output %>% subset(P.value < 0.05 & ngenes >= 3)

write.table(combined_output, quote=FALSE, row.names=FALSE, sep='\t', file=paste0(data_dir, 'tdeg_enrichr.tsv'))






```

Save pseudo-bulk aggregates in pseudotime bin for VAE modeing

This part is the old part

```{r eval=FALSE}

library(scales)
#
# # re-load t-DEGs for this celltype:
# tdegs <- GetDynamicFeatures(seurat_subset, 'tDEGs')
#
#
# # 1: Bin the pseudotime:
# # seurat_subset <- BinPseudotime(seurat_subset, n_bins=50)
#
# # 2 : Bin the expression matrix:
# seurat_subset <- BinFeatureMatrix(
#   seurat_subset,
#   features = tdegs$gene_id,
#   pseudotime_bins = 'pseudotime_bins_50'
# )
# dim(GetFeatureMatrix(seurat_subset, 'pseudotime_bins_50', 'avg_exp'))
#
# # 3: export data for RVAGene:
# outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
# cur_name <- paste0(cur_celltype, '_', cur_group, '_RVAE')
# seurat_subset <- SetupForRVAGene(
#   seurat_subset,
#   pseudotime_bins = "pseudotime_bins_50",
#   feature_name = "avg_exp",
#   name = cur_name,
#   outdir = outdir
# )

```

Save pseudo-bulk aggregates in pseudotime bin for VAE modeing

This part is the new part as of October 2022

```{r eval=FALSE}


o_tdegs <- read.csv('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC_Old_tDEGs.csv')
y_tdegs <- read.csv('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC_Young_tDEGs.csv')

tdegs <- o_tdegs
tdegs <- y_tdegs



library(scales)


# get list of genes to use:
genes_use <- unique(c(tdegs$gene_id, chromvar_motifs))
length(genes_use)

# make sure this isn't the old version of this function!!!
seurat_subset <- BinFeatureMatrix(
  seurat_subset,
  features = genes_use,
  pseudotime_bins = 'pseudotime_bins_50'
)
dim(GetFeatureMatrix(seurat_subset, 'pseudotime_bins_50', 'avg_exp'))

outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
cur_name <- paste0(cur_celltype, '_', cur_group, '_RVAE')
seurat_subset <- SetupForRVAGene(
  seurat_subset,
  pseudotime_bins = "pseudotime_bins_50",
  feature_name = "avg_exp",
  name = cur_name,
  outdir = outdir
)

```

<!--
Save the pseudo-bulk aggregates just for TF genes that were not tDEGs
October 2022
```{r eval=FALSE}




# get list of genes to use:
genes_use <- chromvar_motifs[!(chromvar_motifs %in% tdegs$gene_id)]

# make sure this isn't the old version of this function!!!
seurat_subset <- BinFeatureMatrix(
  seurat_subset,
  features = genes_use,
  pseudotime_bins = 'pseudotime_bins_50'
)

outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
cur_name <- paste0(cur_celltype, '_', cur_group, '_TFs_RVAE')
seurat_subset <- SetupForRVAGene(
  seurat_subset,
  pseudotime_bins = "pseudotime_bins_50",
  feature_name = "avg_exp",
  name = cur_name,
  outdir = outdir
)


``` -->


Plot reconstructed expression heatmap:

```{r eval=FALSE}

library(future.apply)
library(viridis)

# RVAE directory
outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/reconstructed/'
cur_file <- paste0(cur_celltype, '_', cur_group, '_RVAE_recon.csv')
cur_name <- paste0(cur_celltype, '_', cur_group, '_RVAE')


seurat_subset <- LoadRVAGeneOutput(
  seurat_subset,
  pseudotime_bins = 'pseudotime_bins_50',
  name = cur_name,
  rvagene_csv = paste0(outdir, cur_file)
)
print(names(seurat_subset@misc$pseudotime_bins_50))
print(names(seurat_subset@misc$pseudotime_bins_50$feature_matrices))



################################################################################
# Function to order genes by their expression level
################################################################################



seurat_subset <- RankFeatures(
  seurat_subset,
  pseudotime_bins = 'pseudotime_bins_50',
  feature_name = 'ODC_Old_RVAE_scaled',
  dynamic_features = 'tDEGs',
  thresh = 0.75,
  selection = 'median'
)

z_df <- GetZ(seurat_subset, 'pseudotime_bins_50')

feature_mat <- GetFeatureMatrix(seurat_subset, 'pseudotime_bins_50', 'ODC_Old_RVAE_scaled')

# melt matrix for plotting:
plot_df <- reshape2::melt(as.matrix(feature_mat))

p <- plot_df %>% ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  rasterise(geom_tile(), dpi=800) +
  scale_fill_viridis() +
  scale_color_viridis() +
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )


pdf(paste0(fig_dir, 'tDEG_', cur_group, '_heatmap2.pdf'), width=5, height=10)
p
dev.off()


################################################################################
# below: old ###################################################################
################################################################################

scaled <- feature_mat

# order rows by the time they reach 0.75% min expression
exp_thresh <- 0.95
ordering <- future_lapply(1:nrow(scaled), function(i){
  match(names(scaled[i,])[scaled[i,] >= exp_thresh][1], colnames(scaled))
})
ordered <- scaled[rownames(scaled)[order(unlist(ordering))],]
print(head(ordered))


# reverse ordering
ordered <- ordered[rev(rownames(ordered)),]
z_df$rank <- match(rownames(z_df), rownames(ordered))
z_df <- arrange(z_df, rank)

# change colnames of ordered to the bin names:
colnames(ordered) <- levels(seurat_subset$pseudotime_bins_50)

#
# melt matrix for plotting:
plot_df <- reshape2::melt(as.matrix(ordered))

p <- plot_df %>% ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  rasterise(geom_tile(), dpi=800) +
  scale_fill_viridis() +
  scale_color_viridis() +
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )


pdf(paste0(fig_dir, 'tDEG_', cur_group, '_heatmap2.pdf'), width=5, height=10)
p / prop_tp / prop_ct + plot_layout(heights=c(15, 1, 1), guides='collect')
dev.off()


# plot just the t-DEGs that are exclusive to this group:
select_genes <- setdiff(o_tdegs$gene_id, y_tdegs$gene_id)
# select_genes <- setdiff(y_tdegs$gene_id, o_tdegs$gene_id)

# melt matrix for plotting:
plot_df <- reshape2::melt(as.matrix(ordered[rownames(ordered) %in%select_genes,]))

p <- plot_df %>% ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  rasterise(geom_tile(), dpi=800) +
  scale_fill_viridis() +
  scale_color_viridis() +
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank()
  )


pdf(paste0(fig_dir, 'tDEG_', cur_group, '_exclusive_heatmap.pdf'), width=5, height=10)
p / prop_tp / prop_ct + plot_layout(heights=c(15, 1, 1), guides='collect')
dev.off()

################################################################################
# save useful variables
################################################################################

save(z_df, recon_df, ordered, file=paste0(data_dir, cur_celltype, '_', cur_group, '_expression_trajectory_data.rda'))

load(file=paste0(data_dir, cur_celltype, '_', cur_group, '_expression_trajectory_data.rda'))

```







Save pseudo-bulk aggregates for gl-cCREs 


```{r eval=FALSE}


# re-load gl-cCRE table:
link_df <- read.csv(file=paste0(data_dir, cur_celltype, '_', cur_group, '_gl-cCREs.csv'))
#link_df_full <- read.csv(file=paste0(data_dir, cur_celltype, '_', cur_group, '_full_link_table.csv'))
peak1_ranges <- Signac::StringToGRanges(link_df$Peak1, sep=c('-', '-'))
peak1_ranges$peak <- link_df$Peak1

peak2_ranges <- Signac::StringToGRanges(link_df$Peak2, sep=c('-', '-'))
peak2_ranges$peak <- link_df$Peak2



o_tdegs <- read.csv('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC_Old_tDEGs.csv')
y_tdegs <- read.csv('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC_Young_tDEGs.csv')

#tdegs <- o_tdegs
tdegs <- y_tdegs


# which putative enhancers are linked to t-DEGs?
cur_peaks <- subset(link_df, Peak1_nearestGene %in% tdegs$gene) %>% .$Peak2 %>% unique
length(cur_peaks)

library(scales)

# make sure this isn't the old version of this function!!!
seurat_atac_subset <- BinFeatureMatrix(
  seurat_atac_subset,
  features = cur_peaks,
  pseudotime_bins = 'pseudotime_bins_50',
  assay = 'peaks'
)
dim(GetFeatureMatrix(seurat_atac_subset, 'pseudotime_bins_50', 'avg_exp'))

outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/'
cur_name <- paste0(cur_celltype, '_', cur_group, '_RVAE_glcCRE')
seurat_atac_subset <- SetupForRVAGene(
  seurat_atac_subset,
  pseudotime_bins = "pseudotime_bins_50",
  feature_name = "avg_exp",
  name = cur_name,
  outdir = outdir
)

```


Plot reconstructed accessibility

```{r eval=FALSE}

cur_group <- 'Old'


library(future.apply)
library(viridis)

# RVAE directory
outdir = '/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/RVAgene/RVAgene/data/reconstructed/'
cur_file <- paste0(cur_celltype, '_', cur_group, '_RVAE_glcCRE_recon.csv')
cur_name <- paste0(cur_celltype, '_', cur_group, '_RVAE_glcCRE')


# load reconstructed data and z embedding
dataset = cur_name
recon_df <- read.table(paste0(outdir, '/',cur_name,'_recon.csv'), sep=',', row.names=1, header=TRUE)
z_df <- recon_df[,(ncol(recon_df)-1):ncol(recon_df)]
recon_df <- recon_df[,1:(ncol(recon_df)-2)]
z_df$gene <- rownames(z_df)
dim(z_df)

# scale between 0 & 1
range01 <- function(x){
  cur <- recon_df[x,]
  (cur-min(cur))/(max(cur)-min(cur))
}
scaled <- lapply(rownames(recon_df), range01)
scaled <- do.call(rbind, scaled)
rownames(scaled) <- rownames(recon_df)

#scaled <- recon_df
# order rows by the time they reach 0.75% min expression
exp_thresh <- 0.90
ordering <- future_lapply(1:nrow(scaled), function(i){
  match(names(scaled[i,])[scaled[i,] >= exp_thresh][1], colnames(scaled))
})
ordering <- unlist(ordering)
ordered <- scaled[rownames(scaled)[order(ordering)],]

ordered_bins <- ordering[order(ordering)]

z_df_tmp <- z_df

ix <- match(rownames(z_df), rownames(ordered))
z_df$rank <- ix
z_df$bin <- ordered_bins[ix]
z_df <- dplyr::arrange(z_df, rank)
z_df$gene <- factor(as.character(z_df$gene), levels=as.character(z_df$gene))
all.equal(as.character(z_df$gene), rownames(ordered))


# change colnames of ordered to the bin names:
colnames(ordered) <- levels(seurat_subset$pseudotime_bins_50)

# reverse ordering
# ordered <- ordered[rev(rownames(ordered)),]

#
# melt matrix for plotting:
plot_df <- reshape2::melt(as.matrix( ordered[rev(rownames(ordered)),]))

p <- plot_df %>% ggplot(aes(y=Var1, x=as.numeric(Var2), fill=value, color=value)) +
  rasterise(geom_tile(), dpi=400) +
  scale_fill_gradient2(low='white', high='darkorchid4') +
  scale_color_gradient2(low='white', high='darkorchid4') +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) + 
  theme(
    plot.margin = margin(0,0,0,0),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.title.x = element_blank(),
    axis.text.y = element_blank(),
    axis.ticks.y = element_blank(),
    axis.title.y = element_blank(),
    panel.border= element_rect(linewidth=1, color='black', fill=NA)
  )


pdf(paste0(fig_dir, cur_group, '_glcCRE_heatmap.pdf'), width=5, height=10)
p # / prop_tp / prop_ct + plot_layout(heights=c(15, 1, 1), guides='collect')
dev.off()


# save the z_df and the ordered heatmap 
ordered_peaks <- ordered 
z_df_peaks <- z_df 
save(ordered_peaks, z_df_peaks, file=paste0('/dfs7/swaruplab/smorabit/collab/AMRF/analysis/ATAC/data/ODC', '_', cur_group, '_accessibility_trajectory_data.rda'))




```


